{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Structure, Systems, Dataset and Loader Tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Topics:\n",
    "- Interacting with `Structure` class\n",
    "- Interacting `PlinderSystem`\n",
    "- Interacting `PlinderDataset`\n",
    "- Loader\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Setup\n",
    "\n",
    "### Installation\n",
    "\n",
    "`plinder` is available on *PyPI*.\n",
    "\n",
    "```\n",
    "pip install plinder\n",
    "```\n",
    "\n",
    "### Environment variable configuration\n",
    ":::{note}\n",
    "We need to set environment variables to point to the release and iteration of choice.\n",
    "For the sake of demonstration, this will be set to point to a smaller tutorial example\n",
    "dataset, which are `PLINDER_RELEASE=2024-06` and `PLINDER_ITERATION=tutorial`.\n",
    ":::\n",
    ":::{note}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting the configuration\n",
    "\n",
    "At first we get the configuration to check that all parameters are correctly set. \n",
    "In the snippet below, we will check, if the local and remote *PLINDER* paths point to\n",
    "the expected location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "local cache directory: /Users/yusuf/.local/share/plinder/2024-06/v2\n",
      "remote data directory: gs://plinder/2024-06/v2\n"
     ]
    }
   ],
   "source": [
    "import plinder.core.utils.config\n",
    "\n",
    "cfg = plinder.core.get_config()\n",
    "print(f\"local cache directory: {cfg.data.plinder_dir}\")\n",
    "print(f\"remote data directory: {cfg.data.plinder_remote}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data ecosystem overview\n",
    "This tutorial assumes user have downloaded _PLINDER_ dataset before now. While the examples will run without users doing anything, we encourage users to download the data for performance sake. _PLINDER_ data hierarchy is shown below. We have organized this tutorial to follow this same hierarchy from ground up\n",
    "![image](../static/asset/data/plinder_data_hierarchy.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Structure files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After download all files will be store locally at `~/.local/share/plinder/${PLINDER_RELEASE}/${PLINDER_ITERATION}/systems`. The current default is `PLINDER_RELEASE=2024-06` and `PLINDER_ITERATION-v2`\n",
    "\n",
    "There, we have sub-folders that corresponds to each systems. In each sub-folder, we have:\n",
    "- Receptor PDB: receptor.cif and receptor.pdb\n",
    "- Ligand SDF's: `<biounit_instance_id>.<chain_id>.sdf`. For complexes with more than one ligands, all the sdfs are saved separately.\n",
    "- Sequence fasta: sequence.fasta\n",
    "For more information on the file organization, see \"<link-to-dataset-tutorial>\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Structure Python Abstraction\n",
    "To make interacting with our data seamless, {class} class Structure, a pydantic data class that:\n",
    "- Loads all the structure files + smiles \n",
    "- Gets coordinates\n",
    "- Featurizes residues and atoms of associated protein and ligand molecules\n",
    "- Masks molecules to account for resolved vs unresolved part\n",
    "\n",
    "To interact with the example, do the following:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the structure for a given system_id\n",
    "For this purpose we will use `\"1avd__1__1.A__1.C\"` as our example system id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 13:15:56,532 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:15:56,533 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:15:56,645 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:15:56,645 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:15:56,646 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-23 13:15:56,649 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-23 13:15:56,650 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 0.11s\n"
     ]
    }
   ],
   "source": [
    "from plinder.core.structure.structure import Structure\n",
    "from plinder.core import PlinderSystem\n",
    "from pathlib import Path\n",
    "\n",
    "from biotite.sequence.io.fasta import FastaFile\n",
    "\n",
    "input_smiles = \"CC(=O)N[C@@H]1[C@H]([C@@H]([C@H](O[C@H]1O)CO)O)O\" # Need to account for unresolved part of the ligand\n",
    "input_sdf = Path(cfg.data.plinder_dir)/\"systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf\"\n",
    "system_id = \"1avd__1__1.A__1.C\"\n",
    "protein_structure_path = Path(cfg.data.plinder_dir)/\"systems/1avd__1__1.A__1.C/receptor.cif\"\n",
    "input_sequence_path = Path(cfg.data.plinder_dir)/\"systems/1avd__1__1.A__1.C/sequences.fasta\"\n",
    "list_ligand_sdf_and_input_smiles = [\n",
    "    (input_sdf,input_smiles)]\n",
    "\n",
    "input_sequences = {k: v for k, v in FastaFile.read_iter(input_sequence_path)}\n",
    "\n",
    "# Load holo structure\n",
    "holo_struc = PlinderSystem(system_id=system_id).holo_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': FieldInfo(annotation=str, required=True),\n",
       " 'protein_path': FieldInfo(annotation=Path, required=True),\n",
       " 'protein_sequence': FieldInfo(annotation=Union[dict[str, str], NoneType], required=False, default=None),\n",
       " 'ligand_sdfs': FieldInfo(annotation=Union[dict[str, str], NoneType], required=False, default=None),\n",
       " 'ligand_smiles': FieldInfo(annotation=Union[dict[str, str], NoneType], required=False, default=None),\n",
       " 'protein_atom_array': FieldInfo(annotation=Union[AtomArray, NoneType], required=False, default=None),\n",
       " 'ligand_mols': FieldInfo(annotation=Union[dict[str, tuple[Mol, Mol, Mol, tuple[ndarray[Any, dtype[+_ScalarType_co]], ndarray[Any, dtype[+_ScalarType_co]]]]], NoneType], required=False, default=None),\n",
       " 'add_ligand_hydrogens': FieldInfo(annotation=bool, required=False, default=False),\n",
       " 'structure_type': FieldInfo(annotation=str, required=False, default='holo')}"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show fields\n",
    "holo_struc.model_fields"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List fields\n",
    "- We list all fields and their `FieldInfo` to show which ones are required. `id`, `protein_path` and `protein_sequence` are required. Everything else is optionally. Particularly worth mentioning is the decision to make `list_ligand_sdf_and_input_smiles` optional; this is because ligand will not be availbale in apo and predicted structures.\n",
    "- Out of these field `ligand_mols` and `protein_atom_array` is computed within the object if set to default. \n",
    "- `ligand_mols` returns a chain-mapped dictionary of of the form:\n",
    "    ```python\n",
    "    {\n",
    "        \"<instance_id>.<chain_id>\": (\n",
    "            RDKit 2D mol from template SMILES of type `Chem.Mol`,\n",
    "            RDKit mol from template SMILES with random 3D conformer of type `Chem.Mol`,\n",
    "            RDKit mol of solved (holo) ligand structure of type `Chem.Mol`,\n",
    "            paired stacked arrays (template vs holo) mapping atom order by index of type `tuple[NDArray.int_, NDArray.int_]`\n",
    "        )\n",
    "\n",
    "    }\n",
    "    ```\n",
    "- While `protein_atom_array` returns [biotite AtomArray](https://www.biotite-python.org/latest/apidoc/biotite.structure.AtomArray.html) of the receptor protein structure.\n",
    "- `add_ligand_hydrogens` specifies whether to adds hydrogens to ligand\n",
    "- `structure_type`: could be `\"holo\"`, `\"apo\"` or `\"pred\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.C': (<rdkit.Chem.rdchem.Mol at 0x15f2c2110>,\n",
       "  <rdkit.Chem.rdchem.Mol at 0x15f2c26c0>,\n",
       "  <rdkit.Chem.rdchem.Mol at 0x15f2c1380>,\n",
       "  (array([[13,  4,  5,  7,  9, 10,  1,  0,  3,  6,  8, 12, 11,  2]]),\n",
       "   array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13]])))}"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Inspect ligand_mols\n",
    "holo_struc.ligand_mols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.C': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O'}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.ligand_smiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inspect protein_atom_araay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Atom(np.array([31.221, 22.957, 43.101], dtype=float32), chain_id=\"1.A\", res_id=3, ins_code=\"\", res_name=\"LYS\", hetero=False, atom_name=\"N\", element=\"N\")"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_atom_array[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List structure protein properties\n",
    "Show protein related properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein_backbone_mask\n",
      "protein_calpha_coords\n",
      "protein_calpha_mask\n",
      "protein_chain_ordered\n",
      "protein_chains\n",
      "protein_coords\n",
      "protein_n_atoms\n",
      "protein_sequence_from_structure\n",
      "protein_structure_b_factor\n",
      "protein_structure_tokenized_sequence\n",
      "protein_unique_atom_names\n",
      "protein_unique_residue_ids\n",
      "protein_unique_residue_names\n"
     ]
    }
   ],
   "source": [
    "for property in holo_struc.get_properties():\n",
    "    if \"protein\" in property:\n",
    "        print(property)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Protein backbone mask\n",
    "This is a boolean mask that can be used to select backbone atoms from biotite `AtomArray`. The indices of `True` corresponds to backbone indices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False,  True,  True,  True,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "        True,  True,  True, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False,  True,  True,  True, False,  True,\n",
       "        True,  True, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False,  True,  True,  True, False, False,\n",
       "        True,  True,  True, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "        True,  True,  True, False,  True,  True,  True, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False,  True,  True,  True,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False,  True,  True,  True, False, False,  True,  True,\n",
       "        True, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False, False, False, False,  True,  True,  True, False,  True,\n",
       "        True,  True, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "        True,  True, False,  True,  True,  True, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False,  True,  True,  True, False,  True,\n",
       "        True,  True, False, False, False, False, False, False,  True,\n",
       "        True,  True, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False,  True,  True,  True,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False,  True,  True,  True,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False, False,  True,  True,\n",
       "        True, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False,  True,  True,  True, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False])"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_backbone_mask"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Protein Calpha mask\n",
    "This shows the mask of calpha atoms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False,  True, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False,  True, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "        True, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False,  True, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False,  True, False, False, False,  True, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "        True, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "        True, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "        True, False, False, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False])"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_calpha_mask"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein chain ordered\n",
    "This gives a list of protein chains ordered by how they are in the structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1.A']"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_chain_ordered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein chain ordered\n",
    "This gives a list of protein chains ordered by how they are in the structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein chains for all atoms\n",
    "The list of chain IDs in the structure. Order of how they appear not kept."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1.A']"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_chains"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein coordinates\n",
    "This property gets the 3D positions of each of the atoms in protein molecules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([[31.221, 22.957, 43.101],\n",
       "        [31.828, 24.118, 42.476],\n",
       "        [31.979, 23.854, 41.021],\n",
       "        ...,\n",
       "        [34.341, 35.018, 24.674],\n",
       "        [35.484, 35.831, 24.497],\n",
       "        [33.105, 35.742, 24.15 ]], dtype=float32)]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_coords"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get number of atoms of protein molecule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "964"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_n_atoms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein structure atom names\n",
    "Returns all atoms names the same way they appear in the structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['C',\n",
       " 'CA',\n",
       " 'CB',\n",
       " 'CD',\n",
       " 'CD1',\n",
       " 'CD2',\n",
       " 'CE',\n",
       " 'CE1',\n",
       " 'CE2',\n",
       " 'CE3',\n",
       " 'CG',\n",
       " 'CG1',\n",
       " 'CG2',\n",
       " 'CH2',\n",
       " 'CZ',\n",
       " 'CZ2',\n",
       " 'CZ3',\n",
       " 'N',\n",
       " 'ND1',\n",
       " 'ND2',\n",
       " 'NE',\n",
       " 'NE1',\n",
       " 'NE2',\n",
       " 'NH1',\n",
       " 'NH2',\n",
       " 'NZ',\n",
       " 'O',\n",
       " 'OD1',\n",
       " 'OD2',\n",
       " 'OE1',\n",
       " 'OE2',\n",
       " 'OG',\n",
       " 'OG1',\n",
       " 'OH',\n",
       " 'SD',\n",
       " 'SG']"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_unique_atom_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein b-factors\n",
    "Get protein atom occupancies. If not available in a structure, it's set to zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_structure_b_factor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein residue names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ALA',\n",
       " 'ARG',\n",
       " 'ASN',\n",
       " 'ASP',\n",
       " 'CYS',\n",
       " 'GLN',\n",
       " 'GLU',\n",
       " 'GLY',\n",
       " 'HIS',\n",
       " 'ILE',\n",
       " 'LEU',\n",
       " 'LYS',\n",
       " 'MET',\n",
       " 'PHE',\n",
       " 'PRO',\n",
       " 'SER',\n",
       " 'THR',\n",
       " 'TRP',\n",
       " 'TYR',\n",
       " 'VAL']"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_unique_residue_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein residues number\n",
    "Residue number as they appear in structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[3,\n",
       " 4,\n",
       " 5,\n",
       " 6,\n",
       " 7,\n",
       " 8,\n",
       " 9,\n",
       " 10,\n",
       " 11,\n",
       " 12,\n",
       " 13,\n",
       " 14,\n",
       " 15,\n",
       " 16,\n",
       " 17,\n",
       " 18,\n",
       " 19,\n",
       " 20,\n",
       " 21,\n",
       " 22,\n",
       " 23,\n",
       " 24,\n",
       " 25,\n",
       " 26,\n",
       " 27,\n",
       " 28,\n",
       " 29,\n",
       " 30,\n",
       " 31,\n",
       " 32,\n",
       " 33,\n",
       " 34,\n",
       " 35,\n",
       " 36,\n",
       " 37,\n",
       " 38,\n",
       " 39,\n",
       " 40,\n",
       " 41,\n",
       " 42,\n",
       " 43,\n",
       " 44,\n",
       " 45,\n",
       " 46,\n",
       " 47,\n",
       " 48,\n",
       " 49,\n",
       " 50,\n",
       " 51,\n",
       " 52,\n",
       " 53,\n",
       " 54,\n",
       " 55,\n",
       " 56,\n",
       " 57,\n",
       " 58,\n",
       " 59,\n",
       " 60,\n",
       " 61,\n",
       " 62,\n",
       " 63,\n",
       " 64,\n",
       " 65,\n",
       " 66,\n",
       " 67,\n",
       " 68,\n",
       " 69,\n",
       " 70,\n",
       " 71,\n",
       " 72,\n",
       " 73,\n",
       " 74,\n",
       " 75,\n",
       " 76,\n",
       " 77,\n",
       " 78,\n",
       " 79,\n",
       " 80,\n",
       " 81,\n",
       " 82,\n",
       " 83,\n",
       " 84,\n",
       " 85,\n",
       " 86,\n",
       " 87,\n",
       " 88,\n",
       " 89,\n",
       " 90,\n",
       " 91,\n",
       " 92,\n",
       " 93,\n",
       " 94,\n",
       " 95,\n",
       " 96,\n",
       " 97,\n",
       " 98,\n",
       " 99,\n",
       " 100,\n",
       " 101,\n",
       " 102,\n",
       " 103,\n",
       " 104,\n",
       " 105,\n",
       " 106,\n",
       " 107,\n",
       " 108,\n",
       " 109,\n",
       " 110,\n",
       " 111,\n",
       " 112,\n",
       " 113,\n",
       " 114,\n",
       " 115,\n",
       " 116,\n",
       " 117,\n",
       " 118,\n",
       " 119,\n",
       " 120,\n",
       " 121,\n",
       " 122,\n",
       " 123,\n",
       " 124,\n",
       " 125]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_unique_residue_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get sequence from protein structure\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'KCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRT'"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_sequence_from_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE'}"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get tokenized sequence\n",
    "Get tensor of sequence converted to integer-based amino acid token"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([11,  4, 15, 10, 16,  7, 11, 17, 16,  2,  3, 10,  7, 15,  2, 12, 16,  9,\n",
       "         7,  0, 19,  2, 15,  1,  7,  6, 13, 16,  7, 16, 18, 16, 16,  0, 19, 16,\n",
       "         0, 16, 15,  2,  6,  9, 11,  6, 15, 14, 10,  8,  7, 16,  6,  2, 16,  9,\n",
       "         2, 11,  1, 16,  5, 14, 16, 13,  7, 13, 16, 19,  2, 17, 11, 13, 15,  6,\n",
       "        15, 16, 16, 19, 13, 16,  7,  5,  4, 13,  9,  3,  1,  2,  7, 11,  6, 19,\n",
       "        10, 11, 16, 12, 17, 10, 10,  1, 15, 15, 19,  2,  3,  9,  7,  3,  3, 17,\n",
       "        11,  0, 16,  1, 19,  7,  9,  2,  9, 13, 16,  1, 10,  1, 16])"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_structure_tokenized_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Inspect holo sequences\n",
    "Returns a chain-mapped dictionary of sequences from seqres\n",
    "```python\n",
    "{\n",
    "    \"<instance_id>.<chain_id>\": sequence of type `str`\n",
    "\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE'}"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Structure(\n",
       "    (\n",
       "        'id',\n",
       "        '1avd__1__1.A__1.C',\n",
       "    ),\n",
       "    (\n",
       "        'protein_path',\n",
       "        /Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.cif,\n",
       "    ),\n",
       "    (\n",
       "        'protein_sequence',\n",
       "        {\n",
       "            '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_sdfs',\n",
       "        {\n",
       "            '1.C': '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_smiles',\n",
       "        {\n",
       "            '1.C': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'protein_atom_array',\n",
       "        <class 'biotite.structure.AtomArray'> with shape (123,),\n",
       "    ),\n",
       "    (\n",
       "        'ligand_mols',\n",
       "        {\n",
       "            '1.C': (\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x15f2c1fc0>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x15f2c1540>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x15f2c1e00>,\n",
       "                (\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                ),\n",
       "            ),\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'add_ligand_hydrogens',\n",
       "        False,\n",
       "    ),\n",
       "    (\n",
       "        'structure_type',\n",
       "        'holo',\n",
       "    ),\n",
       ")"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.filter(\n",
    "        property=\"atom_name\",\n",
    "        mask=\"CA\",\n",
    "\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List ligand properties\n",
    "Show ligand related properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input_ligand_conformers\n",
      "input_ligand_conformers_coords\n",
      "input_ligand_templates\n",
      "ligand_chain_ordered\n",
      "ligand_template2resolved_atom_order_stacks\n",
      "resolved_ligand_mols\n",
      "resolved_ligand_mols_coords\n"
     ]
    }
   ],
   "source": [
    "for property in holo_struc.get_properties():\n",
    "    if \"ligand\" in property:\n",
    "        print(property)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ":::{todo}\n",
    "\n",
    "The input ligands are provided using dictionaries.\n",
    "These dictionaries contain information for each ligand:\n",
    "- input_ligand_templates: 2D RDKit mols generated from RDKit canonical SMILES (taken from annotation table)\n",
    "- input_ligand_conformers: 3D (random) conformers generated for each input mol\n",
    "- input_ligand_conformers_coords: positional coordintates for 3D conformers\n",
    "- resolved_ligand_mols: RDKit mols of solved (holo) ligand structures\n",
    "- resolved_ligand_mols_coords: positional coordintates for holo ligand structures\n",
    "- ligand_template2resolved_atom_order_stacks: paired stacked arrays (template vs holo) mapping atom order by index\n",
    "- ligand_chain_ordered: ordered list of all ligands by their keys\n",
    "\n",
    ":::"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ligand atom id mapping mapping\n",
    "\n",
    "Unlike the protein sequence - there is no canonical order to ligand atoms in the molecule.\n",
    "It can be further complicated by automorphisms present in the structure due to symmetry, i.e. there is more than one match that is possible between the structures.\n",
    "\n",
    "This is important when calculating ligand structure loss, as the most optimal atom order can change between the different inference results. Typically, it is accepted to take the atom ordering resulting in the best objective score and use that for the loss calculation.\n",
    "\n",
    "Occasionally futher ambiguity arises to to part of the ligand structure being unresolved in the holo structure - this can lead to multiple available matches. We use RascalMCES algorithm from RDKit to provide all the possible matches between the atom order in the input structure (from SMILES) to the resolved holo structure.\n",
    "\n",
    "This is provided as stacks of atom order arrays that reorder the template and holo indices to provide matches. Each stack is a unique order transformation and should be iterated."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ligand conformer to input smiles mapping\n",
    "\n",
    "Each RDKit ligand mol that is generated from SMILES matches the atom order that is in the starting SMILES.\n",
    "This order is retained when 3D conformer is generated.\n",
    "\n",
    "! NOTE: While we ensure that each PLINDER ligand can be loaded and sanitized into RDKit 2D molecule, some ligands may struggle to generate sanitizeable 3D conformers. We nonetheless are trying to provide each ligand with as sensible starting structure as possible and make it accessible via coordinate arrays."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample_system_2 = PlinderSystem(\n",
    "#     system_id=\"102m__1__1.A__1.C\",\n",
    "# )\n",
    "# sample_system_2.holo_structure.ligand_mols['1.C'][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.C': (array([[13,  4,  5,  7,  9, 10,  1,  0,  3,  6,  8, 12, 11,  2]]),\n",
       "  array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13]]))}"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.ligand_template2resolved_atom_order_stacks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below we use RDKit functionality to draw indixes for `'1.C'` ligand 2D, conformer and holo structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit.Chem.Draw import IPythonConsole\n",
    "IPythonConsole.drawOptions.addAtomIndices = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAACWCAIAAADCEh9HAAAABmJLR0QA/wD/AP+gvaeTAAAfWUlEQVR4nO3deVxU9d4H8A8MOwIDsrmBQiyKiIkryjIwMKi4i1ezHstI63Zb1HrsZnWtbHsKbl6t+5Sa+2OuFKgJzoxCCi6YEGKA7CLLgKgMiDAw5/njjCxeRYSZOdPwfb/445wfh/P7zgv98ju/81sMGIYBIYSQ3jLkOgBCCPlzozRKCCF9QmmUEEL6hNIoIYT0CaVRQgjpE0qjhBDSJ5RGST+SnZ0dGxubm5vLdSBEr1AaJf1IbGzsa6+9FhcXx3UgRK9QGiX9yNChQ1955RUvLy+uAyF6hdIo6UdkMtm2bdvooZ6olwFNBiX9x759+0pLS729vefOnct1LER/UBolhJA+oYd6QgjpE0qjpB9QKJCVhcuX0dTUUVhWhubmjtObN1FXp/3QiB6gNEr03eHDGDYMy5fjlVcwbBg2bVKVT52KS5c6LvvgA/zP/3ASIPmzM+I6AEI0KS8Pzz2HhAQIhQCQlYXgYIwcqTolRB2oNUr02oEDEIk6kqafH2JisGcPpzERfUOtUaLXCgrg49OlZPRobNmiOl6/Hvb2quOMDMyfr9XYiL6gNEr0Go+HtrYuJW1tMLr/z376dIwcqTqWy7UaGNEjlEaJXvP0RGpql5LMTHh6qo4nTUJAgOo4MVGrgRE9Qn2jRK8tXozUVCQkqE4vXMDOnVi+nNOYiL6h1ijRa8OH49AhrFyJd96BqSkqKrBpE6ZM4TosoldoMijpjatXr+7evdvV1fXll1/mOpYeUCpRWIjWVnh4dHSM3rmDAQPA46lO2ZH55ubcREj+zKg1Snrj+++/j4qKcnV15TqQx3nnHZSUQCBAaGjH2ySWjU2XU0qgpLeob5T0hkwms7Gx+fjjj5s6T6/UQX/7G6ZPR1oawsLw5ptcR0P0E7VGSW/4+fmVlpYaGhoaGxtzHUu3hg7FsmVYtgwAdDzjkz8t6hslvcEwzKVLl9zd3W1tbbmO5dEUCty7BysrruMgeo7SKNFfly4hJAS+vhAI8OKLcHPjOiCin+ihnugvf3/U1CA9HadO4e5drqMheotao/1RUVGRra1tj57H9+yBpSXmzVOdXrmCgwfx4YcAwDA4fBiJibh7F25u+Otfofsv7gnRAHpT3+9IJJL09PSXXnqppaXl8VefPYuLFztOr1/H4cOq49Wr8c47CAtDTAwUCowbh5wcjUTca8nJyM/nOgii/+ihvt8JCwsDcO7cOaVS2fu75OXh22+Rm4sRIwBAJIJCgXXr8NNPagpTHZKScOgQGAb5+TAz4zoaoreoNdofHT58eNq0aWY9zCx37qCkRPUlk6kKL17EuHGqHMpatAgXLqg/1r6IjUVpKc6coRxKNIpao/3Ozp079+/fHxwcXFtba9++2mY3Dh9GerrquL5elZIqK+Hg0OUyJydUV6OtrWN6pY5wceE6AqLnKI32O8uWLVvGDkfvoeXL8emnquNffsHbbwPAkCGoru5yWWUlnJ11J4empKQkJydHRkYGBgZyHQvRc/RQT3pl4kRcvoyioo6Sgwd1auUkZ2dnHo8nkUi4DoToP2qNkl556im8/joiIrB+PQYNwvHj2L8fZ85wHVYHLy+vjz76iOsoSL9ArdH+obkZ69bB2xsDB2LiROza1dMfDArCpEkdpy4u+MtfVMdffonPP0dyMjZuhJERLl2Cl5fqWzIZAgJoDjvpJ2j4ff+wdCmuX8f//i/c3JCaiuefx0cfISZGI3XduoXwcFy5guHD8ccfMDDQSC3damlpGT169CeffBIdHa392kl/Q2m0HygpgZcXCgowbJiqZO9efPABCgvVX9ft24iIUI3YNzfHCy/gm2/UX0sPFBYWmpmZDRkyhJPaSb9CfaP9QE4Ohg/vyKEAgoNRXIzGRlhaqrmuJUs6Zj0pFNixA2PGYOVKNdfyH+Li4hQKxdChQ5cuXcqWuLu7a7pSQljUN9oP1Nc/mC6trMAwqK9Xf11790IgALsIaWsreDysX6+FV0+rV69+6aWXysvL2dOqqqrm5mZNV0oIi9JoPzBiBMrK0HnqZ1ERLCzg7Kz+uuzsIJHgtdfArnvS0IDGRsTEoKpK/XV1tWXLlvbxsFu2bKmpqdF0jYSwKI32A08/DWNj7NnTUbJxI2bM0NTLHwMDxMbin/+EiwvYnve8PCxYAE02D5ubm+VyufP9Pwzvv//+0KFDNVcdIZ3RK6b+4eRJREdjzhy4uyMlBeXlkEig6USTnY2lS5GdDWtrNDRg3jwcOqShqn7//XdTU1Ov9hFXhGgRtUb13b17OH8eoaHIy0NICMzM8OqryMrSeA4F4OuLM2cwfToUCjCM8vjxhCeahPokxowZY2xs/P3333t7e5uamo4cObL6gbmqhGgMtUb1nViM8HAEBSElBQDy8/HUUzDU4p9PpRIxMcoff/yVx/uLmVncxo3PPPOMuu4tl8ulUmlSUlJSUlJR55mpgI+PT3p6uhVtxEQ0jwY86TupFAAmTwaA+nr4+MDBAaWl0NqOnoaG+OGHjCFDlu/aVV1Wtm7dOh8fHz8/v77cMicn5+jRo2Kx+Ndff21/Iz9w4MDQ0NCJEyfy+fx169bl5OQIBILk5GQ7Ozt1fAxCHo0h+m3SJAZgfvmFYRgmMZEBmKlTOQnk6tWrTz/9NICRI0fKZLIn/XGZTHbgwIEVK1Z0HlHP4/H8/f3Xrl178uRJhULRfnFpaSk7bnTcuHG1tbVq/RyEPIjSqF6rr2eMjBgjI6a+nmEYZvVqBmDef5+7cOpFIpGpqenUqVObm5sfe31ra2tGRsY//vEPf39/w04dEc7Ozs8999yBAwfq6uoe9bNlZWVPPfUUgLFjx9bU1Kj1cxDSBaVRvZaQwADMtGmq07FjGYCRSjmMSKlUvvXWW7a2tgsXLnzUNYWFhd999110dLSNjU176jQ3NxcKhZ9//nlGRoZSqexJXZWVlaNGjQLg5+fXi/YvIT1EaVSvrVrFAMwHHzAMw9TWMoaGjJkZc/cu12Exe/bscXBweO+999pLGhsbT548uXbtWn9//86dTm5ubitWrDhw4IBcLu9FRVVVVaNHjwbg7e1dUVGhvk/wZDHcvn2bYZjr169nZ2dzEgPRKHrFpNfY90sCAQCcOgWlElOnwtyc26AALF261N/fXygUWlpapqSkyOXyixcvtu9UamdnJxQKIyIiRCJRH0fROzk5SSQSoVCYnZ0tEAikUungwYPV8Ql6qqSkJDY2NjIy0tfXNzY21tPTs7CwcM6cOdqMgWgaDXjSXzdvwtERJia4dUs1XPTbb7FhA9at4zCoFStWuLu7R0ZG+vn5FRQUjBo1SqFQAODxeGPHjhUKhUKhMDg42FitAwlqamqEQuHvv//u6ekpkUi0PMHp9OnTjY2NcrncxcVl8uTJq1at2rhxozYDIJpGrVH91d78ZDehY1umoaHcBtXY2DhgwAD2bfuhQ4cUCoWRkdH69ev/+te/2rLT8DXAwcHh9OnTIpHo4sWLgYGBUql0ROc9TbXC0tJSLpc3NjZaqn1VLcI1msWkv06dAu4/0VdWIjcXVlYYP57boH744YclS5Z89tlnAG7evAng+eefX7duneZyKMvW1jY5OXnSpEklJSUCgeCBsfqaU1dXFx8ff/z48cmTJyclJX344YfLly/XTtVEa6g1qr86Nz/Znd2CgrQ36v5hlErlV199ZWhoyA4glUqlABYvXqyd2vl8flJSUmRk5Llz5wQCgUQiYUdEac7t27e3bNlSWVm5detWa2vruLg4jVZHOMP1Oy6iEbcrKpo8PRlra4YdlL58OQMwX33FdVyMQqGor69nGObmzZuGhoampqZ3tTtyoKGhISQkBMDQoUPz8/M1UUVzc/PevXujoqLa0/TEiRPZl/VEL9FDvX46KpWa5+e/EB4OIyMATXl5wP0HfE4ZGRmx89xPnz6tVCoDAgLMtTtywNLS8ujRo6GhoeXl5YGBgTk5OWq8eXp6+jPPPDN+/Pjly5cfPXq0oKCALb9w4UJwcDDbiUH0D6VR/cQ+L4+eMgVAcXGxxdmzolGjMHYs13F1OHXqFAABF5nd0tIyMTFRKBRWV1eHhYVduXKl83dv3bq1cePGc+fO9fyGpaWlb7311rhx40Qi0b59+7Kzs/9z7f2srCyBQFBbW6uGD0B0DdfNYaIRw4cPB/Dbb78xDLNlyxYACxYs4DqoLtj5RWfOnOEqgLt374aHhwNwdHTMyspqL1+9enVNTU1MTExTU1P3d5DJZO+9996UKVMGDhzYzX8xe3t7Dw8PCwsLAKNGjaqsrNTwJyPaRq+Y9FBRUVFJSYmdnR27kBKH7b5Hqa6u/uOPPywtLSdMmMBVDObm5omJiYsWLUpISAgJCUlOTh4/fjyAoUOHbt++vaqqqqamZljnfQDvu3fv3pEjR9hWZ2lp6X9ewOfzHR0dHRwcBg0aFBwcLBAIRo0aJZPJhELhlStX2FkAgwYN0vgnJFrDdR4n6vdA85Odt3P16lVuo+ps7969AKZPn851IExzc/PcuXMB8Pn88+fPs4UNDQ2vv/76A4untLW1SSSS6OhoDw8PI6Mu7Q8+n+/p6Tl16tSFCxdu3rz5ypUrD531X11d7evrC8DLy+vGjRva+HhEK6g1qoc6Nz+vXr1aUVHh5OTk7e3NdVwddKeBbGJisn///iVLlhw5ciQiIuLEiRMGBgZisXj27NkmJibsNdnZ2f/+97/T0tLy8/ObmprQqb3p7OwsEAhCQkJGjRpl8Li9rRwdHSUSSXh4ONtPqv35VERTuM7jRP06Nz83bdoEYOnSpVwH1YWbmxuAjIwMrgNRaW1tZTe4t7a23rFjB1t448aNd999d8KECQMGDLCxsWHbmwsWLOimvdkTdXV1bFfG8OHDi4qK1PchCGcojeobdgTPoEGD2P/n8+fPB7B161au4+rA9ify+fzW1lauY+mgUCiio6MBGBgYrF692tfX19ra2t/ff+nSpXv27CkvL1djXXV1dWw/7CQfn7biYjXemXCCHur1DTvUKTQ01MDAQKlUpqamsqdcx9VBLBYDEAgEPB6P61g6GBkZ7dmzRywWNzQ05OXlnThxQnNrQdna2orF4iVz5vxffb1hcDAkEmh4PhXRKBo3qm86dztmZmbW1ta6urpqfyWObuhOx+gDTExMnJycFArF3//+d02vp2djY3P82DG+jQ3KyhAYiKtXNVod0ShqjeoVpVKZkpIC4Isvvjhz5oyLiwsAoVDIdVxdsGlUpxrIrMrKytzcXCsrq4kTJ2qjPktLHD2K2bMhlSI0FBIJfHy0US9RN0qj+kCpVF6+fFksFh87dqyurg7AtWvXSkpKBgwYAB1r9+Xm5t64ccPR0ZEdfq9TJBIJgKCgIPWudtodS0skJmLOHIjFCA2FWAxfXy1VTdSH0iiXGhoa4uLijI2N16xZ0z68pudu3LjBbtEuFovZ7AnAxMRk/PjxxsbGBQUFN27cAMAup6Qj2K7bsLCwxw4P0j5uehssLJCQgLlzkZyMsDCIxRgzRqsBkD6jNMqlnTt3zp8/v6qq6ujRo+wr9cdqbW09d+4cu0s7O9eTLXdzc2OXjo+IiDAzM2tra6utrQ0LCysoKHj22WdPnjzZ/YRFrdHZjlF0ejun7YrNzZGYiOhoJCQgJARJSeBuchfpDa6HCvRr7777bn19fV5e3qZNm7q/sn2zTHZ5JJaFhQW7WWZOTk77lVVVVW+88cb27duZTltj6sgmw0ql0t7eHkBBQQHXsTyosLAQgJ2dXVtbGzcRNDczc+cyAMPnM/fnU5E/BWqNcmnMmDGpqalVVVVjHvYc19jYmJ6enpiY+PPPP3eeu+3m5hYVFTVr1qzAwEBTU9MHfsrJySkmJiYjIwOAs7OzVCoNCwvLzMwMCgqSSCTcTuXOysqqra11cXFxd3fnMIyHYpuiAoHA0JCj4SsmJti/H4sXIz4eL7+MS5ege/0e5KEojXJp0aJF+/btGzhwYFBQEFvS/rJILBanpqa2b5Zpb28vEAiEQuHMmTPZjYx6yMnJSSqVsltjhoaGSiQSLW+N2Vl7xyhXAXRDJ3ob2Ey6Zg1Wr1bl0Bs3kJYGhQL+/vDyUl126hT8/GBnpzq9ehUGBhg5kpuYCeihXjdUV1cfOHBgxYoVnVuLPB7P399/7dq1v/76a8+fNJuamlatWhUdHS2TydoLZTIZ2+D19PRU74ScJxIVFQVg165dXAXQDR1cwIWJjWXs7JglS5gXXmAGDWJWrGDYGajDhzMpKR2Xvf46s2YNVzEShiaD6ojODUw3N7eXX345Pj6e3WxDXWpqasaOHQvA29vn+vUWNd65e19++WVqairDMAqFwtraGsD169e1VnsPPTCDVidcvMiYmzPtab26mhk6lGH/AlEa1TH0UK8T5s+fX1xcLBKJRCKRh4eHJqqwt7c/derUjBlRhoa7p00zlkrh5qaJero4ffp0Q0NDaWlpYGBgRkZGfX29l5eXDi5rZHLhwkcTJtT5+urQMKyjRzFnTsejuqMjVqxAQgKee47TsMhDUBrVCf/617+0UAufzz9+PEUk4pWWIiwMUim6mSPa1tb25Zdf1tbWzps3b+rUqb2orqmp6ciRI8uXL2d36eBsOFEPPJWY+P7Fi1i5kutAOikuxgMv4jw88NNPquOYGAwYoDouL8d//ZdWYyNd0Zz6/oXP5yUnIyAAJSUIDsb9LdcegsfjvfHGG1OmTCkuLu5dXRUVFUOGDNm+fXtaWhp05B3OQymVSE0F7m9GrSOsrNDY2KWkoQHW1qrjTz5BfLzqKypK+9GRzqg12u/Y2CA5GbNm4dQpBAZCIsGjpmU2NDQUFxf3upPB3d197dq1MpmspqZGLpezabR9TIIOycxEbS1cXbtrnGvfyJHYtatLydmzHb8qJye4uqqOOw0lJpyg1mh/xK6JERaGqiqEhqLrzpgqCoXixIkT3t7eubm5fanL0dHRx8fn/PnzbW1tBgYGcrm8L3fTiFOnAEDXhmEtWYKyMnz2GVpa0NaGH3/E4cN49VWuwyIPQWm0n2JncoeHo7oaYWHIzn7wAmNj4ylTplhZWb399ttPevOWlpaqqioAzc3Nzz77LICgoCALCwuGYWbPnl1TU6OOT6A+bBrVtd4GOztIJJBKwefDxgZxcUhIULVGHRzQeQUGa2tqkHKM66EChEv37jGzZjEAY2vLXLzYy5u070O8cePG/Px8hmH27t07b948tlAul7MHmZmZbP+At7d3RUVFHyNXG4WCsbZmAIa74bSPx9X8VNIz1Brt10xNcfAgZs7ErVuIjHyCtYMzMpCSAgC1tRgxAkolAOTm5iYmJgKYMWPGmjVr2CsH3H+h7Ofnd+bMGV9f39zcXIFAUFFRoeYP0zsXL6K+Ht7eeJK5YdrG1fxU0jP06+nvTE1x5AjmzYO3N+RyrFyJ9vxWWYnVq7tcHBGhevwvLERcHADY28PeHpcuAcCzzz7r6uoKgM/nP3SMFLs15pgxY/Ly8gQCAbuOH8ekUkDH3tGTPxtKo0Q1kzspCbW12LoVb76pKr99Gzt3drnS0xNJSQAQEYGUFDQ3A8CiRWAXTgkICFiwYEH3dTk4OJw+fXrChAn5+fnTpk3r9WgqtdHNjtFOGIaJi4t744032JG87c18ojsojRIAMDaGpSUA+PsjMxO//PLwy0QiVRq1tYW3Ny5cAID338fChU9Ql62tbXJy8qRJk0pKSgQCQVFRUR+D773mZqSlwcAAISGcxfA47E6lJiYmPB7v7bff5mwBKvJo9CshXRgbIy4Of/sbmpoe8t3QUJw/rxoVfuIEAgN7WQufz09KSpo8eXJpaalAIGDX+uTA+fNoasKYMbC35yYAohcojZIHRUVh1Chs2PCQb1la4uBB1QsPPr9PtdjY2IjF4pCQkLKyMoFAcO3atT7drneCgvDHH/j2Ww6qfhInT57My8u7du3aL7/8kpeXx2X7nTyMAXN/FwpCjh3Dp5/i7FmUlWHsWOzYgRdewM2bGqyxsbFx9uzZUqnU2dlZLBb7aGdrzOZmfPUVjh3DrVvw9MSaNdDBuVWdlJeXt7S0DBw4sL6+XqFQODg4WNFAUV1Ck0HJQ7i4YPVqvPeexiuytLRMTEycM2eOWCwOCwsTi8WjR4/WeK3PPAOZDP/8J5ydkZyM2bNx6BB0bBvqztrXxLKxseE2EvJQ9FBPHu6//xv3l97XLAsLi4SEhPDw8Orq6rCwsN9//12z9WVl4cQJHD6MSZPg6oqXXsLbb+PjjzVbKdFrlEZJBxcXzJmjOjYxwdatWLZMG/Wam5snJibOnj1bJpOFhISwG0lpSlYWxo6Fo2NHiUiErCwN1kj0HaVR0uHuXdjadpyOHQsDA7S1aaNqU1PTgwcPzp0799atW+Hh4RfYsVTqVVCA69dRV4cHHo35fMjlUCjUXyPpHyiNkg7Z2Th0qOP07l3ExaG1VUu1m5iY7N+/f/78+bdv346IiDh37pwabnr3LsRivPMOfHzg4YHNm+HqirKyLteUlWHwYBgbq6E60i9RGiU6xMTE5MCBA0uXLr1z505kZGR6enpv7sIw+O03fP45BALw+QgPxxdf4OpV2NnB0BABASgsRFpax/Xff6/L75eI7qM39aSL1lbcvq06rq/nIAAej7dz505DQ8Pdu3eHh4cnJib2cMH8uzKZRXIykpJw8iSqq9tvh4AAiEQQiTB+PHg8APj6a8yciVdewaBBSEpCXp5q9XtCeoXGjZIOW7fi9dfRvsdzWxtKS3HvHkxNtR1JW1vbiy++uHPnTvY9/qO2tm9ra8vMzBSLxYmJiWXFxWXtq6o4OSEoCFFRmDWrS3dvu+xsJCRALoeXFxYtUs2EJaRXKI2SDlu34uBB1ax5ADIZnJy4SaMAlEplTEzM9u3bLSwsfv75Z2Gn5+7CwsLk5OSkpCSpVNq+nL65uXnuvHku48dDJHrkviiEaAA91BMdZWhouG3bNnNz82+//XbmzJlr164NCgoSi8VisfgSuzAfAMDNzU0oFAqFwunTp7evbUqINlEaJbrLwMBg8+bNcrl89+7dGzZsaH9ysrW1FQqFIpFIJBL1Zdf7oqKi/Pz8yMjIy5cvp6amLl682MnJSU2xk36E0ijp4O7eZeFNMzMsXqx6K8MVAwODHTt2FBQUXLhwwc/Pb9asWZGRkRMmTOCpI6zc3NyTJ09GRkZWV1cvXrx4w4YNmzZt6vttSX9DaZR0EAi6pFFra+zbx1009xkaGs6YMSM8PPzDDz9U751nzJhx+vRpAJGRkTk5OS4uLuq9P+knaNwo+RPYvn37tGnTNHf/K1euxMfHr1q1SnNVED1GaZToutLS0ps3b/r6+qr9znv27Glqajp79uzXX39tZGR08OBBtVdB+gMa8ER03auvvnrixIkHVsgvKipycXHh8XhZWVnOzs7Ozs5chUcItUaJrjt37tzgwYM7l6Snp7/55pu3bt1KTU2trKx87bXXWrU285+Q/0BplOi0M2fOFBQU2HfdK2nKlCnsBs7BwcGNjY22trZGRvSylHCG0ijRaZs2baqvrx82bNijLpg/f76Njc3t9oUACNE6+htOdFdra2t2djaAgICAzuU5OTlZWVm7d+8eM2bMhQsXWlpaaG8iwiF6xUR01/79+5ctW2ZhYXH+/HkPDw+uwyHk4eihnuiuik8/ndraam9jM2LECK5jIeSRqDVKdNWdO4yvr7K6utHc3Jq6PokOo9Yo0VXffWdQXs4zNrbuw+IjhGgBpVGiq44dA8PA0BATJ3IdCiHdoTRKdFJZGfLzMWAA5HKMG8d1NIR0h9Io0UmxsZDJAMDCAlOmcB0NId2hNEp0UloaLC3R0ICBA+HlxXU0hHSHht8T3XPhAq5ehVIJAHZ2oK1BiG6j1ijRPZ99BkND3LsHAF1n0xOig6g1SnRMaytOnEBLi+rUwYHTaAh5PGqNEh2zbRuMjVVP9AC8vTmNhpDHozRKdExaGuztYWYGAEZGmDqV64AIeQyaDEp0j0KBs2fx44/IyUF8PHWPEh1HaZToGKUSLS2q1ighfwb0UE90Rl4ewsNhZYXBg+HlhQMHVOXTpiEhoeOyzz7DypWcBEjIQ1EaJbpBLkd4OAICcOcO6uqweTNWrIBUCgD37qGtreNKhaLjPT4hOoDSKNENP/0EGxusXw92V6XwcLz6KjZv5josQh6Pxo0S3ZCbiwkTYGDQUTJxIuLjVceHD+OPP1THv/4KWjqP6BJKo0Q3NDfD2LhLiZmZaiITAGNjmJqqjnk8rQZGyONQGiW6wc0NBw92KcnNhbu76nj2bMybpzpubERxsVZjI6Rb1DdKdENUFM6fR2qq6rSuDt98g+hoTmMipEeoNUp0g4sLtm3DvHkICoKVFaRSTJ+OmBiuwyLk8Wj4PdEldXU4exbNzfDzQ/uOypmZGDYMAweqTsvK0NwM2m+Z6AxKo4QQ0ifUN0oIIX1CaZRwrLGxcdOmTZcuXWJPv/nmm/ZjQv4UKI0SjikUCg8Pj2vXrgFITU0tKysrLCzkOihCngClUcIxPp9vb28PoLm5OT4+fuHChVxHRMiToTRKdEVhYaGtre2uXbvS0tK4joWQJ0DjRgnHGhoakpOT29raIiIiPvjgg/Ly8rq6Oq6DIuQJ0IAnQgjpE3qoJ4SQPqE0SgghfUJplBBC+oTSKCGE9AmlUUII6RNKo4QQ0if/D2Yz2BYHnSFNAAABQXpUWHRyZGtpdFBLTCByZGtpdCAyMDIzLjA5LjUAAHice79v7T0GIOBnQAB+KG5gZGNIANKMzGwMGkCamYUDQjOxM2SAaUY2BRMgg4mRBcJgZAQqAWtByKAKAFVADGVigklwMCiA+BgmcDMwMjAyaTAxMiuwMLMwsbAysLKxMLKyM7BzsDCxczJwcrEwcnEzcPIw8PAy8PKxMPKyMIgwArWysfDycLKziu8DGQP31GfbqQfiBc+ABBmE//seWKCQbA1iK3GaHfgZu8sexDaq4DzwgaVtP4jd8NZ4v6CcKphdqb1i38FdfmA1CyQ27394+S2Yfexdn73n5Vlg9opb++xvxms5gNgu/7gd7J8p2YHYYtUZDglL+cFq/jRMc3Du/wV2wzKBF/YL+5eCzeeI2WPXyXoTzGZ5cGX3zwrPA2C9AIxhR2+Z1+LoAAABp3pUWHRNT0wgcmRraXQgMjAyMy4wOS41AAB4nH1TwY7cMAi95yv4gbHAgA2HHnZmdldVtYnUTvsPvff/VfBoNpnKqu1YDrxgeLwskOP79dvvP/A56nVZAPA/y93hFyPi8gF5gPPr+9cVLreX88Ny2X6utx9AmgtzPmNfbtvHw0JwgZOU1tiowgkLu2mPTwqOsX9bE8mlkjWX4WdpbQJk2OBUi4kwdqAiqu48AQqsCSTp3ghOVFA61llIzbuxhNeE89RE3GdZtjuSA6g9/IZduU+APbOkKNfJPLI0Naw+AVpEpELUWqtxqOjaZ9V4BIxinVAMajBaCWcZUjQo/GSdB4+kjkYzYPaGS+OqlBdqI6JZKdG4DSTY4Sqa1Ys5d5kheSRpFAKyLL86i8wYJ4nLg8fWLXgOZHOVeZqaTKa9ibeUCIkhzqh8Xa9P4rvL8byt112OOeuuuZy8Kyt8ILt+KJQRib69f+FdKmlsux4onj4wtPc+jbZ3mOLx5zg+jHRsG42NnmPRSDlWPbTjjuQD7TQ2OdBLY9N/oundrkfKjgTl++Nfj/PyFwVuzOczflG1AAABR3pUWHRTTUlMRVMgcmRraXQgMjAyMy4wOS41AAB4nE2QsWoEMQxEfyXlLXiNxpJly0vgYFOkyqZJFVKkSBkuhINr7uMj3xG8lTWDNfPsdT08btPL+3o8Pn/gfhy2yYf/sw/rNm03C9vD9TBLVOWKMFNkq7mEZeaYUNUCRSIWzT4uc4pVhBEQJWe7O5BiGmZEkkLsFsVCqQr3NBW5XfNctzyYYqWSuTfAuwzmYTXXvokIqKobiaxDeI2BckiOlxAWL6uFOxOykfMuHJVThjtZAfiOeDsn6eVSjYv0mAomdMZkLDksjqOlWnfUstQbIZGKaedmf1T1hSl8nk/fr7+nn0axjy+n81ekhiHQ0hCp8RDcZAhpeYjcdAhtZYjS6hC12RDWsEOAM+wh0LDDQGrYgYAbdiiQcHlznBws+uf3t16enJVDieX6B3k8kURS7OVYAAAAAElFTkSuQmCC",
      "text/plain": [
       "<rdkit.Chem.rdchem.Mol at 0x15f2c2110>"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.input_ligand_templates[\"1.C\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAACWCAIAAADCEh9HAAAABmJLR0QA/wD/AP+gvaeTAAAgAElEQVR4nO3deVyN6f8/8NepTvteWrSpVFREMbYJgxbGnmXse4hmGj9MZjC2IXwGMXYzFOKrwkg0NAxjpmaaJC1UKsmS1CmiU5069++Pk5bTKeXUOSfez0ePz+Puuq/7vt+3D++57vu67utiMQwDQggh70tO2gEQQkj7RmmUEELEQmmUEELEQmmUEELEQmmUEELEQmmUEELEQmn0/aWlpR0+fFiwXVZWFhQUlJGRId2QCCGSpyDtANqrCxcuREdHX7lyZcGCBVwud/HixXl5eUpKSjY2NtIOjRAiUSwafi+Onj17JiQkCLa3bNliaWn5xRdfSDckQoiE0UM9IYSIhdIoIYSIhdIoIYSIhd6Nvqfk5OQNGzYItvfv3x8QEJCTkwNg5MiRM2fOlGpohBCJojRKCCFioYd6QggRC6VRQggRC6VRQggRC33FJEpxMQoKYGoKZeV65UVFKCyEuTkUFeuVv36NlBS8eQNDQ3TtCjn6jxMhHxH6By/KL7/AxgZvP0+qtWcPbGyQnl5bkp+P6dOhq4u+fTF0KBwdYWyMH38EddwR8tGgNCqG4mK4uiI0FCtW4L//kJmJqCj06IHly+HrK+3gCCESQg/1YlizBunpOH0akydXl1hZwc0NY8di715MngxXV6nGRwiRBGqNvi8eD0FB6NGjNocKyMlh61YAOHJEKnERQiSM0uj7Sk5GSQk++UTErq5doaWF2FiJx0QIkQJ6qG/c559Dof6fT2lp7fbz5wBgZib6WHNz5OS0WWSEEBlCabRxkyfD3LxeybVriI6u3maxADTaI88wNOyJkI8EpdHGzZyJfv3qlfB4tWnUyAgAcnNFH/voETp2BACGqU64hJAPFLWY3peDAzQ1ERcnYte9e3j1Cn37AsDmzejVC8HB4PEkHCAhRDIojb4vBQXMnIk7dxAeXq+cYbBmDQDMnw8AISGIj8esWbC0xMaNePZMCqESQtoSpVExbNgAKytMn47Nm5GRgcJCxMTAywvh4fDxwYABABAbi4MH0a0bnjzB2rUwM8OoUYiOlrXPnO7duzds2DB3d/dvvvlG2rEQ0s5QGhWDjg5u3cLw4VizBra20NdH//74/Xds3ow9e6rraGjA2xt37+K//zBjBuTkcPEi3NzQtSsCA/H6tVRvoJa/v39AQMCVK1fS09Nv374t7XAIaU9o2mZRiouRnw9zc+GpSTgcFBSgUyfhqUlevEBCAl6/hpERXFygpNTomfPyEBSEvXur+6Y0NDBlCpYuRbdubXAbLeDp6blz586uXbv6+fk5OTnNmTNHuvEQ0o5QGpUGHg9nz2LfPty8CYBvYDDSyWn+okWjR49WUJDO2Ik7d+6sWrVKTU2toKBg3rx5M2bMkEoYhLRHlEalKikJ+/Zdz8oacuUKAFNT04ULF86fP99IMJpKGkaPHr1p06bu3btLK4C6bty4oaur261bNwB///337du3e/fu3adPH2nHRUg9lEalr6Sk5NSpUz/99FNSUhIAFotlbGxsa2trZmamq6trYGBgaGhoZGSkr6+vqamppaWlqamprq7eujFcvXo1Li4uMzNTSUlp3759rXvy97N48eKcnJy+ffuuXbs2Kirq8OHDc+fODQgI2Lx5syvN+UJkCaVRGRIfHx8YGBgSElJVVSWygqKioqqqqrKyMpvNVlRUVFNTE2wrKCgoKCgoKysrKSlpaWkZGhrq6+vr6+t37NhRRUVFR0dHR0fHwMCgiTcGpaWlcXFxgvTdZvfXQFkZ5OXBZje2Pzw8PCUlZe3atQD4fL6cnNz+/fu5XO6yZcskFyQh70JfMckQFxeX4ODgbdu2BQUF7d69++nTp0IVKioqKioqWnTOmsyrpKSkqqqqoKDAZrMFmVdZWVlZWVlTU1OQatXV1XV1dVNTU/X19XV0dNhs9vz5883MzJycnGxtbe3s7KytrRWF+tbeT3Y2Nm5ERAQKCgDAwgITJ2LVKujqNnGQnJwcwzBRUVGrV69uhRgIaT3UGpVRVVVVly5d2rVr1/Xr1yXw/xGbzVZXV5eTk5OTk1NRUTEyMsrOzn7x4oVQNWNjYwcHBysrK3t7e8GGpaUlq0Vfu8bHw80NPB7mzYOLCyorcf06QkLQqRNu3ICJSd26dVujANasWcPn83/44Qfx7pWQVkZpVNalpaUdPXp03759JSUlYp5KUVFRU1NTW1tbVVVVVVVVW1tbTU2tY8eOnTt37tq1q7W1tZmZGfvtI/aAAQP+/vtvDw8Pe3v7+/fvp6enP3z4sOHbBg0NjUx39w5KSrCzg50dbG1haws1NdER8Hiwt8eLF/j7b9jb15ZHRGDsWAwfjosX61avSaOlpaWLFy92cHBYuXKlmH8IhLQ6SqPSUV5eLicnV5Oz8vPz9fT05OXlG6v/6tWr4ODgn376KS0trekzq6mpaWhoqKqqCl6JCh7bzc3Nu3Tp4ujoaGxsbGRkJNeM2aemTp166tSpkydPTp06VVDC4/Fyc3OzsrJSUlJSU1Oz3qoyMJDLz693sI4O7O3h4AArK1hZwd4eXbpAXh7nz2PcOGzejFWrhK83eTJCQ3H/Pt6+nN2xY8fNmzf5fP6AAQMMDAzOnDmjoaEBYNiwYd7e3u+MnxCJoTQqaQzDrF69OiwszMfH56uvviorKxs1apSZmVlycvKhQ4d69OjR9OHx8fG7du06depUVVWVtbW1kZGRurq6tra2np6enZ2do6OjqampqampqqrqOyMpKSnhcDimpqYi0/eECRPCw8PDwsK8vLyaOAmHw9FNT8e9e0hPr/7JyEB5uXA9LS0UFcHPD7t3459/REx3ffQo5s7FoUNYsOCdkRMiU6iLSdJYLJaPj4+1tbXgIT00NLR3796bN2+Oi4vbsGHD2bNnmz7cxcXl+PHjAQEBFy5cWLx48XuHERERsX379p49eyYkJERFRTVMu+Xl5QDe2aekK1gVVTCdVY2nT5GaiqwsZGUhJQWpqdDQAItV/e2WhYWIEwmmdn306H1viBCpoTQqBSZ1OlKSkpL69esHwNnZOTU1tflnECeHArh8+fLKlStHjhw5duzY7OxsBwcHoQqCIQFKTXzY2oSOHaunW60hmCdQ8L8iW8qC16ktHIdAiCygNCplgnE8gu2WdXk3oagIZWXQ0xP+9r+OuXPnenl5RUdH6+rq2tft7XlL0Bp9zzTakOAtsGBIU24uGl5R0FDV12+dyxEiQTTDk5R169bt33//BRAfH+/o6Ciihp8fFBSql36qa/58KCjUWx6quBjLl8PYGLq66NgRmprw9ERMjMjr/vrrr/7+/oMGDUpOTk5MTGxYQdAabZ2BojWcnQFAZKM7Obm2AiHtCrVGpWDXrl3nz5/n8XiamppTpkw5ceLErFmz0tLSDh8+LKJ2VRWqqkTMTypUzuHA1RXp6ZgxA8OHQ1UVycnYvRsDB+LUKUyYIHR0RETEH3/8oa2tff/+/bi4uIZdW63cGhXw8sLy5dizB15e9dZWKSnBzz/DxAQDB7bm5QiRDIZInKCLnMPhvH79WlDy4sWLqqoq0bWXLmUA5tkz4fLZsxmAeXuG6l9Pn65XJy+PsbRkNDWZ/Hyho3/55ZchQ4Z89dVXw4YNe/78ecPLCprGSUlJLby5d/n2WwZgpk1j8vKqS9LSGFdXBmD+7/9a+VqESASlUZnXnDRaWMgoKjJDhog4PCSEAZjt2xvu4XK5+Q3Saw0bGxsA6enpYoQuSlUV4+/PsNmMvDxjacmYmjIAo6nJHDnSyhciRFLoof6DEBuLigoMGiRi19ChAHDzJpYvF9oj+Ka+sVM2c8BTi8nJYcsWLF6MqCg8egR5edjYYMSIpj+oJ0SWURptJ8aNE54JKT29dvvxY6CR8ZgGBlBVra7QEmINeHonc3MIfYmUng4lJdG3QIhso576dsLcHJaW9X7qTjnK5wNodPESJSVUVrb0gm3VGhVp/3506YJNmyRxLUJaG7VG24nAQAhNiT9nDjIzq7f19IC3Qy+FlJSgqAg9e7b0gm3bGhXi5gYWCydPIiCg+l4IaT+oNfpBcHEBgKQkEbsEY0J79WrpKSXaGu3cGcOHg8uFyCFfhMg2SqMfBCsr9O2L8HBkZwvv2r4dLBamT68t+e03fP990+fj8/mVlZUsFovd+NT0rezLLwHgp5+qPxglpP2gNPqhCAxEZSXc3XHtWvWr0idPMGcOLlyAj0/tAs7FxfD2xuDBTZ9Mok/0Au7u6NYNT57g/HnJXZSQ1kBp9EPxySe4dAk8HoYOhYYGOnSAqSlOnsSKFQgMrK22aBEmTcJnnwHAP/+I+DgKgISf6GsIJlvZvVuiFyVEbDTfqMxLSkJWFjw9hTviExLw6BFGjkTd2UIrKnDjBlJTUV4OExMMGQJj49q9x48jIADx8VBWRlwcRo1CYiIMDRte88WLFwYGBh06dMgXmo+5TZWWwswMHA7+/Re9e0vuuoSIh9LoR+PxY/TujagoODmhtBTOztiwAZMmNVL3sZmZmYmJyeOWDzgVy4oV+N//MGsWjh2T6HUJEQOl0Y/GjRtISYGPDwAsWYLXrxEUBAAHD4LLhZ9f3bpZWVnW1tZWVlaZmZkADhw4cO3aNQDl5eVTp06dPHlyWwWZkwNraygo4OFD4QFehMgqejf60Rg0qDqHJiYiMrL6FWRmJtauxfDhQnWF3o0uWrTozJkzZ86c0dTUbNuF7C0sMHo0ystx6FAbXoWQVkVp9OPj5IT4eGhpobIS06Zh3TrY2QG4evVqzaOJyJ76J0+eFBYW9mz5SP6WEYx8OnCAZsIn7QWl0Y+S4EuhsjJ4eWHRIgDh4eFLly4tKysT7BfZUx8YGOjr69vmsQ0eDCcnPHuG0NA2vxYhrYHS6EdMXR0rVoDF4vP5W7ZsOX78uIqKimBPw9ZoSUnJzZs3PTw8JBGYIFnv3Nlwz927d1+8eCHYvnfv3vnz5589eyaJkAhpHKVRAjk5uZiYmE8++QTAypUrg4ODG059f/jw4QULFjRngfv3k5+fHxoaGhoaevHiRUyfDgMDxMcjNramApfLnTJlyvTp06OjowGcPHly48aNubm5Hh4ekh5OQEh9lEYJANR89Dl06NDPP/88KSkJwIMHDyrfTg1lY2Mzbdq0tgvg5s2b4eHhRUVFpaWlUFLCvHllysp/hIXVVFBRUQkJCZk6darg1/Hjx584ccLX19fJySkrK6vtAiPknSiNkno8PDz09PQErc6cnBxFRUUHB4eZM2dmZWXdunWrsLCwja775MmTsWPHent7T5o0CQB3yRIrLS333bufPn1aU6fuyqkqKip37twZMWJEeXm5q6trG0VFSHNQGiUi+Pr6enp6ysvLMwyTmpp6/PhxPz8/Nze3Dh06WFtbe3l5rV+//ty5c1lZWa017lhFReXcuXPjxo1bsmQJn89XMTFxHTSIx+Pt37+/sUOcnZ1DQ0P5fL5gTCsh0kLD70lTCgoK7t+/n5qampKSEh8fn5CQUFp3SWdAQ0PD1tbW3t7excXFxcWlR48e6nXnk245Ly+v1atX9+zZ86+//vr00087dOjw6NGjmsVOAgICLCwspkyZkpiY2KlTJy0tLX9/fycnpylTpohzUULEQWmUtEBlZeWjR48EKTU+Pj41NbXhe0ljY2NBSnVwcLC3t7e3t6/7MN6YuLi47t27KygoDBkyZP/+/fb29gA++eSTuLi4o0ePzp49G8DKlStjY2MVFRVnzpzZpUuXTZs2ycvLa2tr7927V1VVtQ1ul5BmoTRKxFJcXJycnCxIqYL0WjP4VEBTU7Nbt26ClOri4uLs7Cwy5YWFhf38889sNnvcuHFz5swRFAYHB8+aNatHjx4JCQmSuJnWVllZeebMmby8vOHDh3ft2lXa4ZC2QmmUtKbKysq0tLSalBofHy80rlNeXt7CwqLmJYCDg4OlpWVjzdWKigoLC4u8vLybN2+2x36k9evXa2tr29jYrFy58ubNm7q0+ukHitIoaVtFRUV1XwIkJycLBqXW0NbWdnBwqHkJ4OLiUvMVAIB169atX79+woQJoe32oyY+n9+vX79z58517NhR2rGQNkFplEhUeXl5SkpKYmLi3bt37969e+fOHQ6HU7cCi8VSUlL69NNPDx06ZGlpmZ+fb25uXllZmZGRYWlpKa2w39vFixe///57V1fXXbt2STsW0lYojRIpe/r0ad2XAKmpqYJyZWXlO3fu2NnZzZgx48SJE998801AQIB0Q30/DMOMHj167dq1vWku6g8UpVEiWwoLC4OCglavXs3lcjt06HD+/HklJaVevXrp6Ojk5uaqqalJO8AWCAsLGzhwoJ6e3siRI3/44QdnZ2dpR0TaBA2/J7JFT09v2bJleXl5o0aNevHixZAhQ9LS0vr27VtUVHTy5ElpR9cyXbt2/frrr728vMaMGUM59ANGrVEic0JCQs6fP3/06NFVq1bt2bOHxWKNHz8+PDzcwcEhKSmpOaNQCZEk+XXr1kk7BkKEGRoaOjg4jBw5UkdH5+rVqykpKWpqak+ePHF1dbWyspJ2dITUQ61RIuuioqImTZpUUlICwM3N7cqVK9KOiJB6KI0S2VJaWpqXlyfYrml4JiYmjhgxQjDbk6+v725ay57IEupiIrIlJibGzc3Nzc1t0aJFNYVOTk7//vuvtrY2gD179qxYsUJ6ATbq7Fls3Yq7d4XLt25FZGT1dmAgIiKEK1RVYccOREW1eYSkjVBrlLQDubm5xsbGpaWlPXr0yM7OBuDp6RkZGdl2s/G/h7Fj8euvcHTE7dt4Ows2ALBYmDmzejVrNhsjR+LcuXoHlpVBRQXTp+P4cYkGTFqLDP0tJKQxe/fuffz4saamZkZGxtChQwFERUV169ZNaBoUqWOzkZIC+l7pY0NplMiSV68waVLNT/n27YJPRQMCAjp16gRAXl4+Ojrax8cHQGpqaqdOnd65pN2KFSsEy0xJgJ4eJk/G+vXIyZHMBYlMoDRKZImyMiZOrPk5WVzs4eHRsMm5d+/enTt3slis58+f29jYJCcnN3a+mJgYDodTIcEl77dtA8Ng8WKJXZBIn4K0AyCkDkVFTJwIgMPh6Orqzp040WTgwJqp7+vy8/OztbUdM2bMmzdvevbsee7cuZEjRwrVKS8vX7ly5dmzZ93c3Fo3zLlz8f33sLAQscvMDMuXY8MGnD+PsWNFVOBwEBNTr0SCSZ60CWqNElmSm4usrLTo6O729hnXriEry6NPn8bqjhgxIiEhQV1dvbKycvTo0TsbrGu/cePGuXPndujQQcyg+HwASE3F6NHVJU+f4u0MKiL4+8PKCl9+idevRey9eRP9+9f7GTxYzACJlFEaJTLkzcyZcHOzW7hwI5//9ahRcHPDqVNN1Hd0dHz48KGJiQnDMMuWLVtc51m6vLw8PDz80qVLkyZNysnJ+d///teiSGreuK5ahcBAAOjYETduVBd26YL79xs9VkUFgYHIzcWmTSL2Dh2K5OR6P7dvtyg0InsYQmRDRUWFnZ3d5cuXBb+Wl5c3/8BevXoJ/j736dOnqqpKqIKTk1NzzlNcXJyamirYNjRknj5lGIbZt49ZuJBporDGmDGMkVG9X5WUmAcPGICZObO6UEGBGTtW+EAulwGY6dObEyORRdQaJeIqKys7fPjw1atXa0oqKyvXrVvXgq82KyoAsNnsoKCgWbNm3bhxA4CiomIzj2az2XFxcYLFQf/55x8bG5vX9R+nDxw40MThsbGxOTk5ABISEmrG/Ne0N+s2PEUWNmb3bigowN+/mTdB2jFKo0Rc69evT0hIuH79ek3Jjz/+mJCQcOvWreaeYv58bNsGoA+fn9CrV/d582BnB1dXrF6NuuOZiorg4YGGK9c/fQoPjxB39zVr1gDIysqysLDIzc2t2d+3b1+hI3g83o4dOwTbQUFBkZGRALp06XL/bXZsOo3a2b07jZqb47vvEB7ezD8C0o5RGiXi2rJlS91e8uzs7OvXr8+dO7dZB1dVAUBAAA4dwuefo3//jrdv6wwciNGjoa+P7dvRpQuuXauuXF6OK1dw757wSd68wZUryMjYsGHDiRMn5OTkOByOtbX1r7/+KlQxNjb22LFjANhs9qZNmwoLCwHY2dmlpaUBMDIyqqioeFuItDQAMDaGtjZKSwHgs8+gqQkAHTti4cLq2Jvw//4funRp1h8DaddowBNpTQzD+Pj4bNu27eHDh++ufewYfvsNJ06gY0csW4YlS9CvH6KjUbMCc2oq3NwwcSJSUmBk1JwApk2bZmZmNmzYMB6PN2HCBB6PB+C7776bMmWKo6Mjl8u9+/aj9yVLlghGpHp6etZEe+nSJcEE+97etR901jQ8J0+uvdD69cKXNjQUHgKlqIiffsKCBagZLNClC0xNhQ+Uk4OdHWi9u3ZM2i9nyYcgMjJy1apVDMOEhITY2Nh88803Xl5e/fv3z8zMFH0An88wDFNeznh6MrNnM3w+4+TEaGgweXnCNcPCGIBZs4ZhGObZMwZgfH2F66SnMwDz7bc1BfHx8SoqKiEhIYJfeTye+PfYung85vTp6j8G0t7RQz0R15s3b16/fs3lcl+/fj148OBTp05NnDhxwIAB9vb2osdspqfD1RUvX0JREeHhePAACxYgMRHu7jA0FK48ZgyUlfHbby0KydnZubS0VNDpBEBBQeaeujZtwhdfwNMTT59KOxQiNpn760XanTNnzsTGxgI4cuSIn5+fsbExAAUFBWtraw0NDREH2Nqib194euLqVair48IFfPYZAIhcP1lBAaamyMioLfnnHwgt2VB/ieZ2oXt36OnhyhX07IkjRzBqVFOVo6OjORzOpEmTJBUdaRmaKI9ICo8HT0/s2wc7OzAMFi5EdjYiIqobm56e2LwZq1aJOLB/f8THo7wceXkwNoa6OrS06lWorMTz5/j2W/zwg2RupVU8f45586qnIp0xA/v2QV1dRLUdO3Y8evQoOzu7YY8ZkRH0UE8khc3GnDlwc0N2Nlgs7N8PPT188QUqK6v7vxubq+npU2hr1/46Zw4eP6738+efkoi/tRkaIiICBw9CVRXHj6N7d/z1l4hqy5Yt+/rrryUeHWkBSqNEgqZPh78/3N3x7Bnk5REcjOHDIS8Pe3vIyeHBAxGHlJbiyRM4Oko8VklgseDtjbg49OyJ7GwMHgx/f/B40g6rvsLCQnpmbRq9GyWS5eOD4mK4u+OPP6Cnh4ULAUBLC4MG4do15ObCzKxe/eBgVFZizBipBCsZ9vaIicHq1dixAzt2IDt785YtX0hlAdSioqKs+lJSUp49e/bw4UMLkfNZEQCURkmzHD6M8nIsXSpcvncv2Gx4e9crTEnBxYvIyQGLBUtLjB4NW9t6Fb79Fq9eYeFChIXVFq5fj8GDMXEizp6tHUL5++/w94e1NebNa/V7kilKSti+HSNGIDDw6pkz312+HBAYGDhnzhwAZWVlL1++5PF4L1++1BJ6KSwGDoeTkZHx4MGDjIwMwcaDBw84ojrrtLS0nj17Rmm0CdTFRJrB2RkvXyIzU7jcxgZqarhzp/rXykr4+uLgQSgrw8EBfD6Sk1FVBT8/bNuGuusmMQwKC6GvX+9sx45h0SKwWPj0U+joIDMTCQmwtMTFi+jaFUB1F5OvL4RWBs3IgK1tu+tiEunly5e+vr7Hjx8HMH78+IMHD/71118nT54EoKOjc/Dgwfc4Z2NtzIY1lZWVraysHBwcrOqwtLRksVhi3teHjdIoaYZmptHly/Hjj1i0CNu2QTDUicPBkiU4fRobNmDNmndfKCcHwcGIj8ebNzA0xODBmDYNKirVe0tKsHo1Bg3C+PH1jioowMaNcHfH55+LdZsyIzQ0dNGiRRwOx9DQ8Oeff/682ffVMGMKNKxJGbMVURolzdCcNJqbCysruLri999R959iZSV690ZaGnJzoacnuZjbuUePHs2aNeuPP/5gsVgLFizYuXOnas03sgCA4uLizMzM5mRMJSUla2trcTJmVVUVl8tVFzkgi9C7UdJqLlxAZSW8vSH0j1NBAd7e8PHB5cuYPl1KwbU/5ubmERERBgYGXC730KFDf/75p5OTU0FBgYGBQVZW1oMHDwoKChoepaGhYWNj07lz586dOws2bGxsDBt+G9YSV69eXb16tYmJia6u7pEjR8Q51YeK0ihpHi4X0dEiCtXUqreTkgDAxkbEsXZ2tRVIs50+fZrL5QrmnE5OTr5Xf2orJSUlExMTe3v7us3Mtngq9/f3j4yMNDIyGjduXFxcXO/evVv3/B8ASqOkeZ49g8iF4Wq6iV69AgBjYxF1TEwAoLi4jUL7UAlmm16xYsX48eOXLFly69YtPT29lJSUkpKSy5cve3p6SiAGweO8kZERgD59+ty9e5fSaEOURknzmJvXLkVUQ/AtvICgI4jDETHjW2EhANGfOopSUVEhmM7OxcXlo+3xiImJiY+P19fX9/LyUlZW/uWXXwAEBQXNnj3b2dlZMjlUoG73CXWliERplDSPggI6dRJRWMPaGgByckR8cZSTAwDNG09eVFQ0ZsyYoUOHcrlcU1NTo+ZNM/rh2bt3L4AFCxbUXV9aUPjll19KLAx5eXllZeXCwkI9Pb2EhIRly5ZJ7NLtCKVR0krc3PDdd4iIEDHq6Nw5sFii3wk0cPDgwSlTptRd4/MjVFBQEB4eLi8v713n04aEhIS4uDg9Pb3JdaePbnsbNmwYNWqUjY2NvLx8n8bXu/6YURolraR3bwwejJ9/xvjxcHevLQ8NRVgYxo0T/papEXfv3uXz+VeuXFFRUTlw4ICmYNaSj4xKcPClvn0jTU071XkC2L17N4B58+bVbZ9KwKhRo4YOHfrq1auP9sngnWjcKGmGZg6/z83FkCHIysL48RgwAHw+btxARAS6dUN0NERO4dzApEmT5s2b5+HhsQFQt7kAAAOlSURBVHPnThaL5efn16p30h5UVaFzZzx8iN9+q/kPUlFRkampaVlZWUZGhlQ+tydNoNYoaQZ7e9Rfsriao2PtJ0YAzMzw33/YsQO//oqoKLBYsLXFDz/gyy9rx0W9S/fu3QXLHZeUlBgYGLRC8O3OxYt4+BCdO2PYsJqytDNntOTkBnt6Ug6VQdQaJbKlqKho7ty5XC5XMNhb6NOdj4K7O65exc6dqGmJ8/mwtWXy85+Gh5s07xUzkSRKo4TIkgcPYGcHZWXk5kJXt7owMhIjR8LCApmZkJeXanxEBJq2mRBZsncv+HxMm1abQwHs2wcAS5dSDpVN1BolRGbweDAxwYsXuH0bPXtWF2ZlwcYGSko0t4vMoi4mQmQGm42EBERE1OZQAPv3g8/H5MmUQ2UWtUYJkWFcLszMUFiIf/8Ffcwuq6g1SojMyM/Hq1dQU6ud4aW0FBMnIi2Ncqgso9YoIdLGMNi3D7t21a6NamaGxYuxfDnYbKlGRpqFeuoJkbZFi7B0KczNceoU/vwToaHo0QPffgsvL/D50g6OvBu1RgmRqogIjB6N6dMRHFxv4YCvvsLu3ThwoHoNaiLDKI0SIlUjRiA6Go8fQ+jL19JSmJjA3ByJiVKKjDQXPdQTIj0Mg1u3YGMjnEMBqKqiVy8kJdGqAbKP0igh0lNSgpISWFiI3mthAYaBqAXliUyhNEqI9AheqTX2iaegm556mWQepVFCpEdDo/orT5EePQIg4nmfyBhKo4RIj5wcXFzw4AFKS4V38flITESnTs2c7ppIEaVRQqRqxgy8eYPdu4XLg4Lw5AlmzZJGTKRlaMATIVLF42HgQMTFYe1aLF6MDh1QXIygIHzzDTp1Qnx88xcOINJCaZQQaXv5Et7eCA0Fw0BFBVwuAHh64uhR0Cpy7QGlUUJkQ04O/voLRUXQ1ESfPs1cSJXIAkqjhBAiFupiIoQQsVAaJYQQsVAaJUQ6YmJipk2bNnv27NTUVGnHQsRCs98TIgUMwxw7dmzr1q0cDmfWrFlxcXHSjoi8P+piIkSa+Hy+vb39/fv3pR0IeX/0UE+INB06dGjMmDHSjoKIhR7qCZGasLCwyMjIsLAwaQdCxEIP9YRIAZ/PX7VqVWxs7NatW9lsdo8ePeQbmy6PyDxqjRIiBVwul8Vi9evX7/z58wAcHBwojbZf1BolhBCxUBcTIYSIhdIoIYSIhdIoIYSIhdIoIYSIhdIoIYSIhdIoIYSI5f8DN1f/e4zP+jcAAAFvelRYdHJka2l0UEtMIHJka2l0IDIwMjMuMDkuNQAAeJx7v2/tPQYg4GdAAH4obmBkUzAA0swsbAoaYJoDQjOxK1iAaEZmNgUTIIOJkQXCYGQEKgFpYWSCy6AKMIKUggSY4BIcCgogPoYJ3AyMQFKDiZFZgYWZhYmFlYGVjYWJlZ2BnYOFiZ2TgZOLhZGLm4GTh4GHl4GXj4WJl4VBhBGolY2Fl4eTnVV8H8gYKGbgP/yjx8HROWK/A9OSvX5v7BxUjNv2J53abZelr+NwWvjB/nieVfbfBZkdJJhP7MvfcWZfwEVpe39v2f3L3znuW3rZcd+c/Em24f4r9v9Rstu3Ptdyf+8dhQMRshf2d2qttrvREL+//4TkgbvP4+2n9f/cr/HszP7jd1fYP3o31+6pm/WBsn1L7D1Pd9lP29NwIN/pvr2k23S7LN8v+xxFJtvfzJ5uv/f3BtvSToV90+pm2bdeTd7P/a9z/97H1+zFAPZ/aHtnRbt1AAAB6HpUWHRNT0wgcmRraXQgMjAyMy4wOS41AAB4nH1US27cMAzdzyl0gQj8iZ9FF5mZNCiKzABt2jt03/ujpJyJPAlQyxZo6ol8Ip99aHX9OH//87e9X3w+HFqD/9wR0X4zABxeWhnt+PT87dJOr4/Hm+d0/XV5/dlw1A017rGPr9eXmwfbqUkXGKzWHqC7iJmWBQ7gtPZSIqmHMxq3B+wgLqwNOqua+QJyuyZQAzikgDZAQRp2Jh/ACyjtkkAQC8PKyAEqsVnhMhZyZG7oCq7Kta4oku600IPd8oynr89f6IbXxG+L5JI7wXAAFRlGCcI3PN7wlpQnfqjPsEbClYD6QBbFRcUrNHal4THq9JxMJmk3zoN+pBIVmrLClORboUR5UgkJzQq9R0Z4Cw1gNuuFQeqVAzw3fSSN1bsMnSDKw2EnN9AyEs4Uu8hUJKQDMeThsyUiVNroFEOG7ZC8VSJFgIMTmeWCoWU4he1jyuxJtkTQZ/GGbrQJNNw/sR1bZPcIsykgc9DUaFdzp123ny7nO41uqj1eL+el2hq0pFmDlwBzrcmSGaaAZNLhpahy6hIM5mP3GJtOX53HfOIeE9OJ+ybinPBTs3BOtGvKhuRd8XFOsisyzmncZ61Pu/xjX7J9ger99ktI+/AP0d/XDBgjYBwAAAF2elRYdFNNSUxFUyByZGtpdCAyMDIzLjA5LjUAAHicTVA9ax1BDPwrKd+D86Kv1cceBsNLkcrnJlVw4SJleCEY3PjHR9rD3FU7o9FoR7rdLo/b9fnX7enpxyvuz2W7Jvh6C9y26zZLuH37vEgT6GzLAzQXMdNC4ABOslILZ0wRG0jKCzRWNfNUNICjFOugCzYm71kGsZjTOEAldhQufYWm4KpcJUWRCdCD3dYdkdcPYNgB52hGCdrFrl79RsJppNaRRdfsUepRuVgY5nduDKlQLkZuSxVEmWpeSGhME4BZhcagamHw7ClTkFIK6ZxbZZ2zLg2IoQQTIUwHRZc+g+fRsHNKGXZ60Cl8rSuqoM1jYM+rSc0FjSiXe4TtlzWHyqLmztfl7f3+5+Xf/e+AVvD5/v67wcCD4KCD0OCD8JCDyOgH6UMPosMOYsMP4iMOEgNPETAznEPgwFMMpIGnIMgDT1FQlo+fOTo+vmdCXnrri7XcmVP6/A+ycajyHI96TAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<rdkit.Chem.rdchem.Mol at 0x15f2c26c0>"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.input_ligand_conformers[\"1.C\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAACWCAIAAADCEh9HAAAABmJLR0QA/wD/AP+gvaeTAAAUp0lEQVR4nO3deVCTd/4H8HcIIWgUFTQMHgQvEBStrMcWqdR69BhRa1ftrGerhVqrdqosaKdqp+7WolvFsc7oWruK7LgwZZSVVSur06JYRS2Xyx1QTq0QrnCEHL8/korbXw8geXgivF/DH0/Ck+/388wwb77fPMdXYjKZQEREXeUgdgFERE83xigRkVUYo0REVmGMEhFZhTFKRGQVxigRkVUYo9RF165dy8rKqqqqOnbs2P3798Uuh0g0jFHqIjc3t7S0NJ1O9+abb544cULscohEwxilLho2bBgAT0/PhISEkJAQscshEg1jlLooNTW1uLg4KSkpOzu7tbVV7HKIRCPhzaBERNbgaJSIyCqMUeqwjAy8/nr7S19fy8b+/ZgxA7NmYc4cZGaKUhqRiBzFLoCeHgYDmpraXzY0AMC//42LF3HlCpyckJmJZcuQkQEnpy40f+7cuebm5ldeeUWhUNioYqLuwBilzmhsRG6uZdtoBICzZ/H225bcnDgRKhWysjB4MNzd4ezcqbZv3749adKkPn362LZkIqExRqkziotx8KBlu60NAKqrMWhQ+w5ubqiuRkgIqqowdChGjcLIkRg1yrzROnq03MPjl9retm1bWlraxYsXX375ZUEPgsi2GKPUGf7+OHzYsp2YCACjRyMvD8HBljdzcqBSQaGAoyPKy1FejpSUx5/+9vnn56emDh8+fNT/Gjt2rIuLS1JSUkVFxfLly7v5mIisxBgl64SFYf58+Ptj7FgcPw6VCj4+KCiAwYDSUqjVKC6GWm3eyJPJdDqdWq1Wq9VPtjF27Nj8/PxXX31VrIMgsgZjlDps8GDMndv+0nzWftQofPUVDh/GgweYNg2nT1t+K5XCywteXk828C6wuqGhuLhY/SPztre3d3cdA5Ht8fJ7IiKr8LpREp/JZLp58+bWrVtPPx7MEj09GKMkvpCQkDVr1igUiilTpohdC1GncVJP4tNoNIOevGqK6KnC0SiJz5yh9fX1sbGx5eXlYpdD1DmMUbILsbGxnp6ecXFxdXV1YtdC1Dmc1JNdqK6udnR0HDBggNiFEHUaR6NkF9zc3MwZmp+fL3YtRJ3DGCV7UVVVNWHChLlz51ZXV4tdC1EncFJP9sJkMqWlpU2dOlUikYhdC1EnMEbJ7phMJiYpPUU4qSf78v777//ud78TuwqiTuBolOzL1atXvb29lUql4D21tcFkan9Qv9EIB4ef2Sb6LYxR6n1qarBuHRoaIJFALsexY3B3x/DhKCuz7LBgAXbtQkCAqFXSU4P/csnuXLp0af/+/QJ2sG0b5szBpUv4+mu8+iq2bBGwL+oF+LxRsjseHh7Crsh04QKioy3bq1fjgw8AwGjE9euWNzUaAXunHocxSnZnwoQJwnbQ3Ay53LItlcJkgtEIvR7nzlnefPBA2AKoZ+GknuxRdXV15o9L3ldVVQUFBVnb4s2bCAiAefEST08UFlreLyuDmxscHODkhD//2fIzbpy13VFvwhgle9SvX7/Kykrz9t///vfFixd3vS2TCR9+iEWL8P332LgRACIjERqKO3eQkYG1axEZaYuSqffipJ7skVwuf/HFFwHU1dVJpdKbN2/m5eX5+Ph0uqHKSqxahatX0dICANevIzkZf/gDlEqcOgWjEX/6E2bPBoCwsPZPLVoEd3fbHAn1ArzgiexRc3Pzhg0bCgoK1q5du3r16rKyshEjRnS6lSNHEBVlmcj37QsHB8hk2LoV27fbvGDqzRijZHcuX778zjvv5OXlSaVSg8EwZsyY8PDwN954QyaTdbSJR48wbx5yciCXw2SCXo/Bg7FtG9atg+P/zMAyMzO//fbbd99994svvjAYDKGhobY/Hurp+N0o2ZGysrLXXntt9uzZeXl5/v7+y5YtGzFiRGFhYVhY2NixYw8dOtTU1PSbjXz99df7/PyMd++ipQV9+yIwEOfP4949vP32TzIUwMSJE/V6/f3796VSqaura05OjjBHRj0ZY5Tsgl6vj46O9vPzS0hIUCgUO3fuvHXrVmxsbFFR0YkTJ/z8/O7du7dx40aVSrVr166ampqfbaStrW39+vUrVqyIfPQoS6HA6tW4fh3nz2PmzF/v3TzsNRgMUqlUgIOjns5EJLaUlBR/f3/zH+T8+fPv3bv3kx0MBkNiYuLvf/978z79+vXbtGlTaWnpk/sUFBQEBQU5OztPmjRp69atGo2mI12r1ep9+/YVFhbGxMR8+eWXtjoi6lUYoySqH34wvfnmvmefBTBmzJgLFy78+u4pKSnz5883P0bPyclp5cqV2dnZJpNp9+7dAwYMmD59ekxMjMFg6JbSiSx4iolEYjLhiy8QGYnqar1K9cmaNeGRkc7Ozh356O3bt/fs2ZOQkGA0GqVS6ejRo93d3fft2zdt2jShqyb6/xijJIbMTKxfj9RUAJg1C59/Dl/fTjVQUlKi1Wq3b9+emJgok8l0Op0gdRJ1AE8xkW0cOHAgMTHxxo0bkb9+U1BTE3btwtSpSE2FhwdOnMDly53N0LKysqKiotOnT//zn/90dHQ0GAyNjY1WVU9kBcYo2cbChQtNJtP06dM9PDza362tRUICYmIs97BfuQJvb3z0EQwGbN6M3FysWtWFvoYPH67X6wcNGuTs7Dx+/Hij0fjf//7XRsdB1GmMUbKNxsZGrVbb0tLS3NxsmWIXF+P551FSgrY2rF6Nf/wD7u54+BABAbh+HQcOwMWla32VlJTo9fr6+nr8+Dio7Oxs2x0KUefwnnqyjYaGBg8Pj6qqqilTptTW1iqVSuzejchIvP46AISEYOpUqNW4ehVTpli5RIeXl5eTk9O8efMAjB8/HsDdu3dtcRBEXcEYJdsIDAw0b3h5eVneSk/Hzp2W7SFD4OqKigrY6GT60KFDzRvmC06zsrJs0ixRF3BST4JxckJbW/tLna79Ycm2w0k9iY4xSoIJDMTFi5ZttRomE4YMsXknKpXKxcWlsrLy0aNHNm+cqCM4qSfBbNuGhQuRkwNXV/zrX/j8cyE6kUgkfn5+33333d27d4ODg4XogujXcTRKghk8GN98gzfewOzZuHgRzz8vUD+c15O4OBolITk6QqnE/PlwdrY8PlkAjFESF2OUBObujkePYDCgpQUdu2W+sxijJC5O6klgMhlGjoTR2L4Yp62ZYzQrK4sPiCBRMEZJeOal6PLyBGre3d19yJAhdXV15eXlAnVB9CsYoyQ8gWMUTwxIheuC6JcwRkl4wseo+V4mfj1KomCMkvCEj1HeWU8iYoyS8Dippx6NT7+n7qD39m5QKJwuXlQolUK039DQMGDAALlc3tjYyNU9qZtxNErdIcjV1TU9/Y5gA9L+/ft7enq2tLQUFRUJ1AXRL2GMUnfw8fEBkCdAjF64cCEqKio9PZ0X4ZNYGKPUHYSL0ZdeemnOnDkVFRWMURILY5S6g3Axmp+ff+fOnZdfflkmkwE4deqUzbsg+nWMUeoO48aNgzAxWlhYqFAodu/evX//fgBKpfLGjRslJSWor0ddHQDwJCoJjDFK3WHMmDFSqVStVtt2QXmDwVBVVRUZGbljxw6tVuvm5hYVFZWUlFRQUIAjRxARgYICTJ8OAOfPY9YsvPAC5szB9esAEBWF+HhLQykpeP99GxZGvQqf8ETdQS6Xq1QqtVqtVqvNI1PrJScnb9myJTMzE8D48eN37ty5ZMkSPF4V6sMPsXcvYmLw3HMoKkJ4OK5cwZAhKC7GvHn47jvU10OrtbTV0gKNxiZVUS/E0Sh1E/PXo9988431TV27dm3mzJlz587NzMxUqVRHjhzJyMgwZ2i75GTMmIHYWKxcicRELF9uWcJk5EjMm4fkZACorUVFBSoqUFNjfVXUa3E0St1kyJAhMplsy5YtLS0tGzdudOjSGstZWVkff/xxfHy8ucEtW7a899578p9dKa9fPxgMiI7GM88gLg4jR7b/SqnEw4cAEBeHW7cAoKoKI0Z0oR4icDRKQjt48OCePXsKCgr27t3r5eWl1Wrfe++9mTNnZmRkdKqdvLy8pUuXTpo0KT4+vn///hEREUVFRRERET+foWZSKebPBwCVCk9ell9QYEnV0FCcOoVTpxAR0ZVjIwLAGCWh1dXVhYaGXrp0SalU5ufnJyYmjhgx4tq1a1OmTNm8eXN9ff2TO8fFxR09evQnLZSVlYWFhU2YMCE+Pr5Pnz6bNm0qKiras2dP//79O1rEsmU4cwYpKdDpcO4cMjPx4os2OToiMEZJaK2trUVFRcOGDTO/DAkJycnJiYiIAHDw4MFx48adPHny8YMdFi1apNfrH3+2pqYmMjLS29v76NGjEokkNDS0oKAgOjp6SGcXah44EOfPIzYWISFITkZyMmQyTJgALy/LDu7umDrV2kOlXstEJKSioqIzZ84YjcafvJ+enj5jxgzzH2FwcHB2drbJZKqurt67d6/RaGxsbNyzZ8/AgQMBSCSSJUuW5Ofni1E+0W/jE55INCaTKSYmJjw8/OHDhzKZbP369S+99FJdXZ1Go/n4448rKysBzJkz59NPPw0ICBC7WKJfxBglkdXU1Gzbtu3YsWNGo3HQoEGtra1NTU0AAgMDP/nkk5kzZ9qwL41GExMT4+XltWDBAhs2S70cvxslkbm6uh45ciQtLW3atGn19fVNTU0qlSouLu7q1au2zVAA165de/bZZ1NSUmzbLPVyHI2SvWhra1u/fn1paWlSUpKjoyBXNBsMhjNnzvznP/85fPiwEO1T78TRKNkLmUy2YMECnU4nUIYC0Gq1Go1m5cqVArVPvRPvYiI74ujoqFarhWvfxcVl3bp1wrVPvRNHo2RHJBJJY2Njc3Oz2IUQdQJjlOyIo6NjY2Pj999/L3YhRJ3AGCU7IpVKdTrd+fPnxS6EqBMYo2RHzGsjc7l5erowRsmOODg4SCTSe/dqxS6EqBN4pp7sSEuLb58+P+TmKpqa0Lev2NUQdQxHo2RHFIrBOt0gwOn2bbFLIeowxijZEYkEEgmkUvAkEz1FGKNkRxwcYDLBaMTdu2KXQtRhjFGyIw4OkErR3IzKSrFLIeowxijZEfNoFMCDB2hsFLsaoo5hjJIdcXCAecHQqirLkp1E9o8xSnZEIoH56U46Hc8y0VODMUoia21tfbzd0ACjEc7OcHdHQYGIRRF1Ah/bTGLasWOHXC5fsWKFSqUC0NiIM2fg44Nhw+DuDqlU7PqIOoCjUeq6v/3tb6GhoeXl5V1uobW1tbi42MXFxfyyogK3b+OZZzB0KGpqcOAAAHzwQfv+0dH44QeraiayOcYodd1bb73l6+v7eA36LjCZTK+88srjZ5FUVuL4cXz2GQDU1yMpCQBOnGjf/+xZ1PKGe7IzjFHquszMTCuXPt68ebOXl9eTS9ctWoS4OJSUWFsbUbfho0mo6yQSSXBwsDUtDBs27CeDWScnfPIJNm60zOgBVFfjhRcs2+np1vRGJAiORqkrdDrdsmXLzAvK29y8eejTxzKjB+DmhsuXLT/PPCNEh0RWYYxSVxw+fDguLi4oKGjXrl16vd76BktKsH9/+8vPPrN8Q0pk/xij1BUbNmzYuXOnyWT66KOPAgMD8/LyrGntq68weTLefx9Xr1reGT4c775r2XZyat9TJoNEYk1XRLbH60ap665fv75q1arCwsI+ffrs3LkzPDzcwaFz/5hbWhARgYMHAWDhQhw/DldXQUolEg5jlKxSX18fHh5+9OhRAHPnzv3yyy87fv1Tbi5efx0ZGZDL8emn2LSJI016KjFGyQYSEhLCwsIePXo0cODAQ4cOLV++/Dc/cvIk3nkHWi18fHD6NM8d0VOM342SDSxevPju3bsLFiyora1dsWLF0qVLa2pqHv923759x48ff3wmqqGhYcWKFUePFmm1WLkSt24xQ+npxhgl21AqlWfPnj1x4kS/fv3i4+MnT5585coVAEaj8caNGxqNpqWlBcCtW7cCAgJiY2MLC1+MjdWdPIl+/cQuncg6jFGypVWrVmVmZgYFBd2/f3/27NlhYWEtLS0+Pj5KpfL+/fvR0dEzZswoLCwMCAhISTn/xz86/XaLRHaP342S7en1+r/+9a87duzQ6XR+fn5/+ctfZDLZ4cOHk5KSJBLJxo0bo6Ki5HK52GUS2QZjlISSlpa2ePHisrIyqVTq7Oys1WqlUmlcXNzixYvFLo3IljipJ6FMnTp10aJFAAwGg/kdpVLJDKWehzFKAsrPzwewffv2t956C4C/v7/YFRHZHmOUBJSbmwtgzZo1Q4cOBTBu3DixKyKyPcYoCUWr1ZaWljo5OY0cOdKcp4xR6pEYoySU3Nxck8nk7e3t6OiYk5MDwNfXV+yiiGyPMUpCMY9AzdH55DZRD8On35NQ/B88OBQYOCAg4MGDBxqNZuDAge7u7mIXRWR7jFESysTU1ImpqdiwwZST0zZs2A9BQWJXRCQITupJMDk5AODrK8nJcSwv91AoxC6ISBCMURKGXo+iIkgk8PZGbi4A8DQ99VCMURKGWo3WVnh6QqGwxCjPL1EPxRglYfw4o2/f5miUeijGKAnj8Qi0sRFlZZDL4eUlcklEwmCMkjDkcoweDT8/5ObCZIK3Nxx5WQj1THxQHglGp8PDh3B2xpUrALBkidgFEQmCo1ESxvbtCA7G9u2YPRtOTsxQ6sE4zyIBJCUhOxupqZBIUFODadMQFAQ3N7HLIhIER6MkgORkLFtmWXXe1RWBgbh1S+yaiITCGCUBaLXo27f9pUKBxkbxqiESFmOUBODri/T09pd37sDPT7xqiITF70ZJAGvX4rnn4OmJiRMRF4exY3kLE/VgvOCJhFFbi5MnUV6OyZOxdCkcOO+hHosxSkRkFY4RiIiswu9GycZaW1ujoqKWLl1aWlqal5cXFBQ0adIksYsiEhBHo2Rjcrk8ODi4qalp1qxZZWVljryVnno6xigJRaPRhIWF3eKF99TTcaRAtpebm9u/f3+TyZSRkfHaa6+JXQ6RsHimnojIKpzUExFZhTFKRGQVxigRkVUYo0REVmGMEhFZhTFKRGSV/wPr3jq7g2TBbgAAAVV6VFh0cmRraXRQS0wgcmRraXQgMjAyMy4wOS41AAB4nHu/b+09BiDgZ0AAPihuYGRjSADSTExsDClAmpGZEcJggjEYURgwtRpAmpkFwmdkZmfIAPGZGDkgAki0AkgDggvWx8TNwMjAACSYGBg5WBiZmBmYOFmYmFkYmLlYGFlYWZhYuBlYeRjY2BnYOBTYeDWYRIDKGdgYuFmAVohvAhkFxQx8IckCTuI3jzoWZ0xx1I8Vdiq0PeTol9TlOO2whNP578cd95R0OQbtEHcK47viKLKuy/FqhYjTGpnrjv2Hpji6dAs5LZv9yLH5wFTHg6f4na4v3uD45W+bI9d1bqeLStscJ0vUOe75IuzUf2ibo/LtTkfJFmmnbZ+POu6b3+BovFnGSXHDTUd70y5Hww5+pySBi45PL0x2LLTldeJheuI47/1MxwxBfie2xuWOs2Z2OYoBADHoU95HFAmgAAABoHpUWHRNT0wgcmRraXQgMjAyMy4wOS41AAB4nH1Ty24CMQy88xX5gii28/KhB1goqioWqaX9h977/6qdB9kg1N09xGbGGY8N2GVn9Pk4vv/8mvtDx53k3T8fM5tvcs7tLkYP5nA6v61mue0PPbNcv9bbpwGvn9N3xu5v10vPgFkMRes4SUmD3iYCPUG2IZTTnYwVmj00KGZXoMki5wI1y+v5BTqBlJAtcCcwx0YglwYBO8ErIVmmUhejDRWmBApPbgiVAJAbIWfu6vkZIdYeIhUhmC1hbIQEPLWbFBos+1obrWvtyi3JT9CsUGkvugZNseoOImg2kc2qArjqFGj2oZtIOEHBmav6lxP32XjfFLiUZywoli2kUg2lGkCvm+KMRcWqNGquAVEzwac0Y6liy1IUw4Krddmif9DrK7Y1j04Gmfv4MEzY03qc1rAu5uG6Hsdi6otj+UoCx27Jj4bG5mjIj9sBgvNj9hpC3QgaYwZhhzFKDcXOqRJLNujybJngSnZoApUot9DGaZAL8sZNjXnjGKoAv7Vla4LG/Z8t590fVI/L5SZwMYQAAAE3elRYdFNNSUxFUyByZGtpdCAyMDIzLjA5LjUAAHicTZCxagQxDER/JeUdGGFJtiV5CQQ2RarbNKlCihQpw4ZwcM19fCRDWFfWjEfSs9f19LidL+/r08sHrts4T+t29iKq/wK3h/uJC1hLRCBNElZQxYUrWNEwsyZsoFLCiwRlMHZTgKku3MDUIqilhklMYWrBRAVoBMk0vGwSnjAmVKh1rFHlRA2QOcwisrAAooapPjmSNvY0dkr1rS1MQYt2n+VezT7SgApFt42JNcuAZG82QKmJnARxAElbWAFtQJo/P5JZwlSxQVlKPDyLntPndf9+/d1/eoYoL/v1C6TjIVqnQyB3PpT2Ml31eojc23SDXQ5Vuh6idpty1HGi4I4TBuaOE4hHJxBLtzdn8W/OHrw9+z7/IHLc+x+PwYw6m/aWBQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<rdkit.Chem.rdchem.Mol at 0x15f2c1380>"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.resolved_ligand_mols[\"1.C\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ligand conformer coordinates\n",
    "\n",
    "As you can tell, the input 2D and 3D conformer indices match, but the resolved ligand is different.\n",
    "Thus to perform a correct comparison for their coordinates one should use atom order stacks.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_atom_order_stack, holo_atom_order_stack = holo_struc.ligand_template2resolved_atom_order_stacks[\"1.C\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[36.097, 24.731, 18.551],\n",
       "       [36.841, 24.28 , 17.298],\n",
       "       [38.191, 24.996, 17.307],\n",
       "       [37.93 , 26.507, 17.335],\n",
       "       [37.118, 26.889, 18.595],\n",
       "       [36.636, 28.326, 18.719],\n",
       "       [35.948, 22.08 , 16.874],\n",
       "       [34.96 , 22.767, 15.881],\n",
       "       [36.989, 22.845, 17.232],\n",
       "       [38.879, 24.744, 16.078],\n",
       "       [39.175, 27.211, 17.276],\n",
       "       [35.883, 26.133, 18.477],\n",
       "       [35.31 , 28.501, 19.242],\n",
       "       [35.767, 20.938, 17.325]])"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.resolved_ligand_mols_coords['1.C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[[ 4.40536655, -0.84477619, -0.08008242],\n",
       "        [ 2.98317284, -1.04843568,  0.36677844],\n",
       "        [ 2.69039401, -1.75060402,  1.32850255],\n",
       "        [ 2.0479714 , -0.3906486 , -0.39984463],\n",
       "        [ 0.60866258, -0.61443037, -0.18938695],\n",
       "        [-0.18928392,  0.07150194, -1.31492126],\n",
       "        [-0.18568033, -0.72432991, -2.51346127],\n",
       "        [-1.62589545,  0.3343089 , -0.87305978],\n",
       "        [-2.40286605,  0.87463173, -1.94969441],\n",
       "        [-1.60077386,  1.31926812,  0.30846316],\n",
       "        [-2.92620201,  1.28706237,  1.08432877],\n",
       "        [-4.02302091,  1.74421486,  0.29545668],\n",
       "        [-0.47715314,  1.14905565,  1.1829788 ],\n",
       "        [ 0.08641765, -0.15677436,  1.20698805]]])"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.input_ligand_conformers_coords['1.C'][holo_atom_order_stack]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Interacting with PLINDER systems\n",
    "PlinderSystem is the next layer of abstraction above `Structure`. It provides encapsulation around all structures associated with a particular `system_id`. With these, we can access the `holo` and alternate (`apo` and `pred`) structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_system = PlinderSystem(\n",
    "    system_id=\"1avd__1__1.A__1.C\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check holo structure\n",
    "Since having `holo` structure is a defining feature of _PLINDER_ system, holo structures is by definition available for all systems "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 13:15:58,362 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:15:58,362 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:15:58,477 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:15:58,477 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:15:58,478 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-23 13:15:58,481 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-23 13:15:58,481 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 0.12s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Structure(\n",
       "    (\n",
       "        'id',\n",
       "        '1avd__1__1.A__1.C',\n",
       "    ),\n",
       "    (\n",
       "        'protein_path',\n",
       "        /Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.cif,\n",
       "    ),\n",
       "    (\n",
       "        'protein_sequence',\n",
       "        {\n",
       "            '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_sdfs',\n",
       "        {\n",
       "            '1.C': '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_smiles',\n",
       "        {\n",
       "            '1.C': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'protein_atom_array',\n",
       "        <class 'biotite.structure.AtomArray'> with shape (964,),\n",
       "    ),\n",
       "    (\n",
       "        'ligand_mols',\n",
       "        {\n",
       "            '1.C': (\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x15e875a80>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x15f2c2340>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x15f2c1930>,\n",
       "                (\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                ),\n",
       "            ),\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'add_ligand_hydrogens',\n",
       "        False,\n",
       "    ),\n",
       "    (\n",
       "        'structure_type',\n",
       "        'holo',\n",
       "    ),\n",
       ")"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.holo_structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get annotations \n",
    "This `system` property returns `json` data of annotations for a the system in question. To get the annotations of all other systems sharing the same PDB entry ids, use `.entry` property."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pdb_id': '1avd',\n",
       " 'biounit_id': '1',\n",
       " 'ligands': [{'pdb_id': '1avd',\n",
       "   'biounit_id': '1',\n",
       "   'asym_id': 'C',\n",
       "   'instance': 1,\n",
       "   'ccd_code': 'NAG',\n",
       "   'plip_type': 'SMALLMOLECULE',\n",
       "   'bird_id': '',\n",
       "   'centroid': [36.85636520385742, 25.090288162231445, 17.591215133666992],\n",
       "   'smiles': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "   'resolved_smiles': 'OC[C@H]1O[CH][C@@H]([C@H]([C@@H]1O)O)NC(=O)C',\n",
       "   'residue_numbers': [1],\n",
       "   'rdkit_canonical_smiles': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "   'molecular_weight': 221.0899372,\n",
       "   'crippen_clogp': -3.077599999999999,\n",
       "   'num_rot_bonds': 2,\n",
       "   'num_hbd': 5,\n",
       "   'num_hba': 6,\n",
       "   'num_rings': 1,\n",
       "   'num_heavy_atoms': 15,\n",
       "   'is_covalent': True,\n",
       "   'covalent_linkages': ['17:ASN:A:17:ND2__600:NAG:C:.:C1'],\n",
       "   'neighboring_residues': {'1.A': [9, 11, 15, 16, 17, 34, 35, 36, 123]},\n",
       "   'neighboring_ligands': [],\n",
       "   'interacting_residues': {'1.A': [34, 15]},\n",
       "   'interacting_ligands': [],\n",
       "   'interactions': {'1.A': {'15': ['type:hydrogen_bonds__protisdon:False__sidechain:False'],\n",
       "     '34': ['type:hydrophobic_contacts']}},\n",
       "   'neighboring_residue_threshold': 6.0,\n",
       "   'neighboring_ligand_threshold': 4.0,\n",
       "   'num_neighboring_ppi_atoms_within_4A_of_gap': 0,\n",
       "   'num_neighboring_ppi_atoms_within_8A_of_gap': 0,\n",
       "   'num_missing_ppi_interface_residues': 0,\n",
       "   'num_pli_atoms_within_4A_of_gap': 0,\n",
       "   'num_pli_atoms_within_8A_of_gap': 0,\n",
       "   'num_missing_pli_interface_residues': 0,\n",
       "   'num_resolved_heavy_atoms': 14,\n",
       "   'num_unresolved_heavy_atoms': 1,\n",
       "   'tpsa': 119.25000000000001,\n",
       "   'qed': 0.33723889506423077,\n",
       "   'is_ion': False,\n",
       "   'is_lipinski': True,\n",
       "   'is_fragment': False,\n",
       "   'is_oligo': False,\n",
       "   'is_cofactor': False,\n",
       "   'in_artifact_list': False,\n",
       "   'is_artifact': False,\n",
       "   'is_other': False,\n",
       "   'is_invalid': False,\n",
       "   'posebusters_result': {'mol_pred_loaded': True,\n",
       "    'mol_true_loaded': True,\n",
       "    'mol_cond_loaded': True,\n",
       "    'sanitization': True,\n",
       "    'all_atoms_connected': True,\n",
       "    'molecular_formula': True,\n",
       "    'molecular_bonds': True,\n",
       "    'double_bond_stereochemistry': True,\n",
       "    'tetrahedral_chirality': True,\n",
       "    'bond_lengths': True,\n",
       "    'bond_angles': True,\n",
       "    'internal_steric_clash': True,\n",
       "    'aromatic_ring_flatness': True,\n",
       "    'double_bond_flatness': True,\n",
       "    'internal_energy': True,\n",
       "    'protein-ligand_maximum_distance': True,\n",
       "    'minimum_distance_to_organic_cofactors': True,\n",
       "    'minimum_distance_to_inorganic_cofactors': True,\n",
       "    'minimum_distance_to_waters': True,\n",
       "    'volume_overlap_with_protein': True,\n",
       "    'volume_overlap_with_organic_cofactors': True,\n",
       "    'volume_overlap_with_inorganic_cofactors': True,\n",
       "    'volume_overlap_with_waters': True,\n",
       "    'rmsd_≤_2å': True,\n",
       "    'passes_valence_checks': True,\n",
       "    'passes_kekulization': True,\n",
       "    'inchi_crystal_valid': True,\n",
       "    'inchi_docked_valid': True,\n",
       "    'inchi_crystal': 'InChI=1S/C8H15NO5/c1-4(11)9-5-3-14-6(2-10)8(13)7(5)12/h5-8,10,12-13H,2-3H2,1H3,(H,9,11)/t5-,6+,7+,8+/m0/s1',\n",
       "    'inchi_docked': 'InChI=1S/C8H15NO5/c1-4(11)9-5-3-14-6(2-10)8(13)7(5)12/h5-8,10,12-13H,2-3H2,1H3,(H,9,11)/t5-,6+,7+,8+/m0/s1',\n",
       "    'inchi_overall': True,\n",
       "    'inchi_version': True,\n",
       "    'stereochemistry_preserved': True,\n",
       "    'hydrogens': True,\n",
       "    'net_charge': True,\n",
       "    'protons': True,\n",
       "    'stereo_sp3': True,\n",
       "    'stereo_sp3_inverted': True,\n",
       "    'stereo_type': True,\n",
       "    'number_bonds': 14,\n",
       "    'shortest_bond_relative_length': 0.9492650185663352,\n",
       "    'longest_bond_relative_length': 1.0429011482666541,\n",
       "    'number_angles': 19,\n",
       "    'most_extreme_relative_angle': 1.0452636472719188,\n",
       "    'number_noncov_pairs': 58,\n",
       "    'shortest_noncovalent_relative_distance': 0.7386533194416193,\n",
       "    'number_valid_bonds': 14,\n",
       "    'number_valid_angles': 19,\n",
       "    'number_valid_noncov_pairs': 58,\n",
       "    'aromatic_ring_maximum_distance_from_plane': None,\n",
       "    'double_bond_maximum_distance_from_plane': None,\n",
       "    'ensemble_avg_energy': 32.10719429562964,\n",
       "    'mol_pred_energy': 81.78525064878713,\n",
       "    'energy_ratio': 2.54725622848707,\n",
       "    'smallest_distance_protein': 1.4676218859093153,\n",
       "    'num_pairwise_clashes_protein': 2,\n",
       "    'most_extreme_protein_atom_id_protein': 111,\n",
       "    'most_extreme_ligand_element_protein': 'C',\n",
       "    'most_extreme_protein_element_protein': 'N',\n",
       "    'most_extreme_ligand_vdw_protein': 1.7,\n",
       "    'most_extreme_protein_vdw_protein': 1.6,\n",
       "    'most_extreme_sum_radii_protein': 3.3,\n",
       "    'most_extreme_distance_protein': 1.4676218859093153,\n",
       "    'most_extreme_sum_radii_scaled_protein': 3.3,\n",
       "    'most_extreme_relative_distance_protein': 0.44473390482100467,\n",
       "    'most_extreme_clash_protein': True,\n",
       "    'smallest_distance_organic_cofactors': None,\n",
       "    'smallest_distance_inorganic_cofactors': None,\n",
       "    'smallest_distance_waters': None,\n",
       "    'volume_overlap_protein': 0.052308591466978374,\n",
       "    'volume_overlap_organic_cofactors': None,\n",
       "    'volume_overlap_inorganic_cofactors': None,\n",
       "    'volume_overlap_waters': None},\n",
       "   'unique_ccd_code': 'HSR',\n",
       "   'waters': {},\n",
       "   'crystal_contacts': {'B,125': [1946, 1947]}}],\n",
       " 'ligand_validation': {'num_residues': 1,\n",
       "  'num_processed_residues': 1,\n",
       "  'percent_processed_residues': 100.0,\n",
       "  'average_rsr': 0.192,\n",
       "  'average_rsrz': nan,\n",
       "  'average_rscc': 0.851,\n",
       "  'average_occupancy': 0.8571428571428571,\n",
       "  'percent_rsr_under_threshold': 100.0,\n",
       "  'percent_rscc_over_threshold': 100.0,\n",
       "  'percent_occupancy_over_threshold': 0.0,\n",
       "  'average_b_factor': 70.71357142857143,\n",
       "  'unknown_residue_count': 0,\n",
       "  'atom_count': 14,\n",
       "  'heavy_atom_count': 14,\n",
       "  'num_unresolved_heavy_atoms': 1,\n",
       "  'max_alt_count': 1,\n",
       "  'percent_outliers': {'geometry': 0.0,\n",
       "   'density': 0.0,\n",
       "   'chirality': 0.0,\n",
       "   'clashes': 0.0}},\n",
       " 'pocket_validation': {'num_residues': 9,\n",
       "  'num_processed_residues': 9,\n",
       "  'percent_processed_residues': 100.0,\n",
       "  'average_rsr': 0.12877777777777777,\n",
       "  'average_rsrz': -0.6723333333333333,\n",
       "  'average_rscc': 0.9612222222222222,\n",
       "  'average_occupancy': 0.9367283950617283,\n",
       "  'percent_rsr_under_threshold': 100.0,\n",
       "  'percent_rscc_over_threshold': 100.0,\n",
       "  'percent_occupancy_over_threshold': 77.77777777777777,\n",
       "  'average_b_factor': 31.81567680776014,\n",
       "  'unknown_residue_count': 0,\n",
       "  'atom_count': 61,\n",
       "  'heavy_atom_count': 61,\n",
       "  'num_unresolved_heavy_atoms': 0,\n",
       "  'max_alt_count': 1,\n",
       "  'percent_outliers': {'geometry': 0.0,\n",
       "   'density': 0.0,\n",
       "   'chirality': 0.0,\n",
       "   'clashes': 0.0}},\n",
       " 'pass_criteria': False}"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.system"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get paths of the underlying structure files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`archive` points to the subfolder where all the files (except `apo` and `pred` files) relating to a given system are stored"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C')"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.archive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly, the  `system.cif`, `receptor.cif`, `receptor.pdb`, `sequence.fasta`, and ligand sdfs can be accessed via `.system_cif`, `.receptor_cif`, `.receptor_pdb`, `.sequence_fasta` and `.ligand_sdf` properties respectively. \n",
    "\n",
    "To get all the paths of the structures together, use `.structures`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.cif',\n",
       " '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.pdb',\n",
       " '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/chain_mapping.json',\n",
       " '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/sequences.fasta',\n",
       " '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/system.cif',\n",
       " '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf']"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/system.cif'"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.system_cif"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get binding site water (`.water_mapping`)\n",
    "This returns the information about binding site water "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_system.water_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chain mapping \n",
    ":::{todo}\n",
    "Confirm with Jay\n",
    ":::\n",
    "\n",
    "`.chain_mapping` maps chain ids in system (`<instance_id>.<asym_id>`) to PDB author chain ids "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.A': 'A'}"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.chain_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "linked_structures, linked_archive, get_linked_structure \n",
    "linked_archive, linked_structures,  get_linked_structure, alt_structures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Linked apo and predicted structures.\n",
    "The following properties provides different kind of information about linked structures as described below:\n",
    "- `.linked_archive`: returns paths the local subfolder where the linked structures are saved; \n",
    "- `.linked_structures`: returns the dataframe of linked structures along with all their metrics while \n",
    "- `.get_linked_structure`:  gives the path to a specific linked structure.\n",
    "- `.best_linked_structures_paths`: Gives the best linked structures based on `scrmsd_wave` which is average symmetry-corrected RMSD across mapped ligands weighted by number of atoms. This selects maximum of two alternate structure with at most one `apo` and `pred` each when available\n",
    "- `.alt_structures`: returns the dictionary`Structure` object of the best `apo` and `pred` which the corresponding `holo` chain as key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 13:15:59,036 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:15:59,036 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures')"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.linked_archive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 13:15:59,321 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.10s\n",
      "2024-09-23 13:15:59,651 | plinder.core.scores.links.query_links:24 | INFO : runtime succeeded: 0.58s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_system_id</th>\n",
       "      <th>id</th>\n",
       "      <th>pocket_fident</th>\n",
       "      <th>pocket_lddt</th>\n",
       "      <th>protein_fident_qcov_weighted_sum</th>\n",
       "      <th>protein_fident_weighted_sum</th>\n",
       "      <th>protein_lddt_weighted_sum</th>\n",
       "      <th>target_id</th>\n",
       "      <th>sort_score</th>\n",
       "      <th>receptor_file</th>\n",
       "      <th>...</th>\n",
       "      <th>posebusters_most_extreme_ligand_element_waters</th>\n",
       "      <th>posebusters_most_extreme_protein_element_waters</th>\n",
       "      <th>posebusters_most_extreme_ligand_vdw_waters</th>\n",
       "      <th>posebusters_most_extreme_protein_vdw_waters</th>\n",
       "      <th>posebusters_most_extreme_sum_radii_waters</th>\n",
       "      <th>posebusters_most_extreme_distance_waters</th>\n",
       "      <th>posebusters_most_extreme_sum_radii_scaled_waters</th>\n",
       "      <th>posebusters_most_extreme_relative_distance_waters</th>\n",
       "      <th>posebusters_most_extreme_clash_waters</th>\n",
       "      <th>kind</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1avd__1__1.A__1.C</td>\n",
       "      <td>1vyo_B</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>1vyo</td>\n",
       "      <td>1.48</td>\n",
       "      <td>/plinder/2024-06/assignments/apo/1avd__1__1.A_...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>apo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1avd__1__1.A__1.C</td>\n",
       "      <td>1vyo_A</td>\n",
       "      <td>100.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>1vyo</td>\n",
       "      <td>1.48</td>\n",
       "      <td>/plinder/2024-06/assignments/apo/1avd__1__1.A_...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>apo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1avd__1__1.A__1.C</td>\n",
       "      <td>1nqn_A</td>\n",
       "      <td>100.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>1nqn</td>\n",
       "      <td>1.80</td>\n",
       "      <td>/plinder/2024-06/assignments/apo/1avd__1__1.A_...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>apo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1avd__1__1.A__1.C</td>\n",
       "      <td>1rav_B</td>\n",
       "      <td>100.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>1rav</td>\n",
       "      <td>2.20</td>\n",
       "      <td>/plinder/2024-06/assignments/apo/1avd__1__1.A_...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>apo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1avd__1__1.A__1.C</td>\n",
       "      <td>1rav_A</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>1rav</td>\n",
       "      <td>2.20</td>\n",
       "      <td>/plinder/2024-06/assignments/apo/1avd__1__1.A_...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>apo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1avd__1__1.A__1.C</td>\n",
       "      <td>P02701_A</td>\n",
       "      <td>100.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>P02701</td>\n",
       "      <td>91.22</td>\n",
       "      <td>/plinder/2024-06/assignments/pred/1avd__1__1.A...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pred</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  reference_system_id        id  pocket_fident  pocket_lddt  \\\n",
       "0   1avd__1__1.A__1.C    1vyo_B          100.0         83.0   \n",
       "1   1avd__1__1.A__1.C    1vyo_A          100.0         82.0   \n",
       "2   1avd__1__1.A__1.C    1nqn_A          100.0         76.0   \n",
       "3   1avd__1__1.A__1.C    1rav_B          100.0         81.0   \n",
       "4   1avd__1__1.A__1.C    1rav_A          100.0         83.0   \n",
       "5   1avd__1__1.A__1.C  P02701_A          100.0         94.0   \n",
       "\n",
       "   protein_fident_qcov_weighted_sum  protein_fident_weighted_sum  \\\n",
       "0                              98.0                         98.0   \n",
       "1                              98.0                         98.0   \n",
       "2                              94.0                         99.0   \n",
       "3                              99.0                         99.0   \n",
       "4                              99.0                         99.0   \n",
       "5                              98.0                         98.0   \n",
       "\n",
       "   protein_lddt_weighted_sum target_id  sort_score  \\\n",
       "0                       93.0      1vyo        1.48   \n",
       "1                       93.0      1vyo        1.48   \n",
       "2                       88.0      1nqn        1.80   \n",
       "3                       90.0      1rav        2.20   \n",
       "4                       90.0      1rav        2.20   \n",
       "5                       93.0    P02701       91.22   \n",
       "\n",
       "                                       receptor_file  ...  \\\n",
       "0  /plinder/2024-06/assignments/apo/1avd__1__1.A_...  ...   \n",
       "1  /plinder/2024-06/assignments/apo/1avd__1__1.A_...  ...   \n",
       "2  /plinder/2024-06/assignments/apo/1avd__1__1.A_...  ...   \n",
       "3  /plinder/2024-06/assignments/apo/1avd__1__1.A_...  ...   \n",
       "4  /plinder/2024-06/assignments/apo/1avd__1__1.A_...  ...   \n",
       "5  /plinder/2024-06/assignments/pred/1avd__1__1.A...  ...   \n",
       "\n",
       "  posebusters_most_extreme_ligand_element_waters  \\\n",
       "0                                           None   \n",
       "1                                           None   \n",
       "2                                           None   \n",
       "3                                           None   \n",
       "4                                           None   \n",
       "5                                           None   \n",
       "\n",
       "   posebusters_most_extreme_protein_element_waters  \\\n",
       "0                                             None   \n",
       "1                                             None   \n",
       "2                                             None   \n",
       "3                                             None   \n",
       "4                                             None   \n",
       "5                                             None   \n",
       "\n",
       "   posebusters_most_extreme_ligand_vdw_waters  \\\n",
       "0                                         NaN   \n",
       "1                                         NaN   \n",
       "2                                         NaN   \n",
       "3                                         NaN   \n",
       "4                                         NaN   \n",
       "5                                         NaN   \n",
       "\n",
       "   posebusters_most_extreme_protein_vdw_waters  \\\n",
       "0                                          NaN   \n",
       "1                                          NaN   \n",
       "2                                          NaN   \n",
       "3                                          NaN   \n",
       "4                                          NaN   \n",
       "5                                          NaN   \n",
       "\n",
       "   posebusters_most_extreme_sum_radii_waters  \\\n",
       "0                                        NaN   \n",
       "1                                        NaN   \n",
       "2                                        NaN   \n",
       "3                                        NaN   \n",
       "4                                        NaN   \n",
       "5                                        NaN   \n",
       "\n",
       "   posebusters_most_extreme_distance_waters  \\\n",
       "0                                       NaN   \n",
       "1                                       NaN   \n",
       "2                                       NaN   \n",
       "3                                       NaN   \n",
       "4                                       NaN   \n",
       "5                                       NaN   \n",
       "\n",
       "   posebusters_most_extreme_sum_radii_scaled_waters  \\\n",
       "0                                               NaN   \n",
       "1                                               NaN   \n",
       "2                                               NaN   \n",
       "3                                               NaN   \n",
       "4                                               NaN   \n",
       "5                                               NaN   \n",
       "\n",
       "   posebusters_most_extreme_relative_distance_waters  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "5                                                NaN   \n",
       "\n",
       "   posebusters_most_extreme_clash_waters  kind  \n",
       "0                                    NaN   apo  \n",
       "1                                    NaN   apo  \n",
       "2                                    NaN   apo  \n",
       "3                                    NaN   apo  \n",
       "4                                    NaN   apo  \n",
       "5                                    NaN  pred  \n",
       "\n",
       "[6 rows x 137 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.linked_structures\n",
    "# TODO AttributeError: 'PlinderSystem' object has no attribute 'linked_structure'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1vyo_B.cif'"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.get_linked_structure(link_kind=\"apo\", link_id='1vyo_B')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1vyo_B': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1vyo_B',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1vyo_B.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (958,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " '1vyo_A': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1vyo_A',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1vyo_A.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (947,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " '1nqn_A': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1nqn_A',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1nqn_A.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (928,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " '1rav_B': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1rav_B',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1rav_B.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (976,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " '1rav_A': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1rav_A',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1rav_A.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (976,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " 'P02701_A': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         'P02701_A',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/P02701_A.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (1179,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'pred',\n",
       "     ),\n",
       " )}"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.alternate_structures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get `Openstructure` entities and views\n",
    "`.receptor_entity` returns receptor `mol.EntityHandle` object\n",
    "`.ligand_views` returns `mol.ResidueView` for all ligands\n",
    "\n",
    ":::{note}\n",
    "You must have Openstructure installed to use this property\n",
    ":::\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Others properties\n",
    "This includes:\n",
    "- `num_ligands`: Number of ligand chains\n",
    "- `smiles`: Ligand smiles dictionary\n",
    "- `num_proteins`: Number of protein chains\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Interacting with the PLINDER dataset\n",
    "`PlinderDataset` provides an interface to interact with _PLINDER_ data as a dataset. It is a subclass of `torch.utils.data.Dataset`, as such subclassing it and extending should be familiar to most users. Flexibility and general applicability is our top concern when designing this interface and `PlinderDataset` allows users to not only define their own split but to also bring their own featurizer.\n",
    "It can be initialized with the following parameters\n",
    "```\n",
    "Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame | None\n",
    "        the split to use\n",
    "    split : str\n",
    "        the split to sample from\n",
    "    split_parquet_path : str | Path, default=None\n",
    "        split parquet file\n",
    "    input_structure_priority : str, default=\"apo\"\n",
    "        Which alternate structure to proritize\n",
    "    featurizer: Callable[\n",
    "            [Structure, int], dict[str, torch.Tensor]\n",
    "    ] = structure_featurizer,\n",
    "        Transformation to turn structure to input tensors\n",
    "    padding_value : int\n",
    "        Value for padding uneven array\n",
    "    **kwargs : Any\n",
    "        Any other keyword args\n",
    "``` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plinder.core.loader import PlinderDataset\n",
    "from plinder.core.loader.dataset import get_torch_loader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Interacting with the PLINDER dataset\n",
    "`PlinderDataset` provides an interface to interact with _PLINDER_'s underlying data as a dataset. It is a subclass of `torch.utils.data.Dataset`, as such subclassing it and extending should be familiar to most users. Since, flexibility and general applicability is our top concern when designing this interface, it allows not only allow you to filter the underlying data the way you want but you can also specify your own featurizer.\n",
    "It can be initialized with the following parameters\n",
    "```\n",
    "Parameters\n",
    "    ----------\n",
    "    split : str\n",
    "    filters: FILTER\n",
    "    featurizer: Callable[\n",
    "            [Structure, int], dict[str, torch.Tensor]\n",
    "    ] = structure_featurizer,\n",
    "        Transformation to turn structure to input tensors\n",
    "    **kwargs : Any\n",
    "        Any other keyword args\n",
    "```\n",
    "For example on how to write your own featurizer see [Featurizer Example](https://github.com/plinder-org/plinder/blob/revamp_loader/src/plinder/core/loader/featurizer.py). The signature is shown below:\n",
    "```\n",
    "def structure_featurizer(\n",
    "    structure: Structure, pad_value: int = -100\n",
    "    ) -> dict[str, Any]:\n",
    "```\n",
    "The input is a `Structure` object and return dictionary of padded tensor features."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is where you may want to load a `train` dataset, but for the purposes of demonstration - we will start with `val` due to smaller memory footprint. We are also going to filter for only proteins with single chain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 13:15:59,987 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.08s\n",
      "2024-09-23 13:16:00,250 | plinder.core.split.utils.get_split:24 | INFO : runtime succeeded: 0.00s\n"
     ]
    }
   ],
   "source": [
    "val_dataset = PlinderDataset(\n",
    "    split=\"val\",\n",
    "    filters=[(\"system_num_protein_chains\", \"==\", 1)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's select an item from the newly created dataset for inspection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 13:16:00,641 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:00,642 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:00,749 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:00,749 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:00,750 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-23 13:16:00,755 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-23 13:16:00,755 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 0.11s\n",
      "2024-09-23 13:16:00,877 | plinder.core.structure.atoms:121 | INFO : generate_conformer: MMFFOptimizeMolecule - more iterations are required, extending by 500 steps\n",
      "2024-09-23 13:16:01,328 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.10s\n",
      "2024-09-23 13:16:01,768 | plinder.core.scores.links.query_links:24 | INFO : runtime succeeded: 0.72s\n",
      "2024-09-23 13:16:02,043 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:02,043 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n"
     ]
    }
   ],
   "source": [
    "sample_vata_data_point = val_dataset[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'system_id': '6p7e__2__1.B__1.Y',\n",
       " 'holo_structure': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '6p7e__2__1.B__1.Y',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/systems/6p7e__2__1.B__1.Y/receptor.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.B': 'MIVSAIAANALLESVTKFHCGVIYDYSTAEYVSYRPSDFGAYLDALEAEVARGGLIVFHNGHKYDVPALTKLAKLQLNREFHLPRENCIDTLVLSRLIHSNLKDTDMGLLRSGKLPGKRFGSHALEAWGYRLGEMKGEYKDDFKRMLEEQGEEYVDGMEWWNFNEEMMDYNVQDVVVTKALLEKLLSDKHYFPPEIDFTDVGYTTFWSESLEAVDIEHRAAWLLAKQERNGFPFDTKAIEELYVELAARRSELLRKLTETFGSWYQPKGGTEMFCHPRTGKPLPKYPRIKTPKVGGIFKKPKNKAQREGREPCELDTREYVAGAPYTPVEHVVFNPSSRDHIQKKLQEAGWVPTKYTDKGAPVVDDEVLEGVRVDDPEKQAAIDLIKEYLMIQKRIGQSAEGDKAWLRYVAEDGKIHGSVNPNGAVTGRATHAFPNLAQIPGVRSPYGEQCRAAFGAEHHLDGITGKPWVQAGIDASGLELRCLAHFMARFDNGEYAHEILNGDIHTKNQIAAELPTRDNAKTFIYGFLYGAGDEKIGQIVGAGKERGKELKKKFLENTPAIAALRESIQQTLVESSQWVAGEQQVKWKRRWIKGLDGRKVHVRSPHAALNTLLQSAGALICKLWIIKTEEMLVEKGLKHGWDGDFAYMAWVHDEIQVGCRTEEIAQVVIETAQEAMRWVGDHWNFRCLLDTEGKMGPNWAICH',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         {\n",
       "             '1.Y': '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/6p7e__2__1.B__1.Y/ligand_files/1.Y.sdf',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         {\n",
       "             '1.Y': 'Cc1cn([C@H]2C[C@H](O)[C@@H](CO[P@](=O)(O)O[P@@](=O)(O)OP(=O)(O)O)O2)c(=O)[nH]c1=O',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (5555,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         {\n",
       "             '1.Y': (\n",
       "                 <rdkit.Chem.rdchem.Mol object at 0x15e8c2880>,\n",
       "                 <rdkit.Chem.rdchem.Mol object at 0x15e8c0900>,\n",
       "                 <rdkit.Chem.rdchem.Mol object at 0x1bf6b89e0>,\n",
       "                 (\n",
       "                     <class 'numpy.ndarray'> with shape (2, 29),\n",
       "                     <class 'numpy.ndarray'> with shape (2, 29),\n",
       "                 ),\n",
       "             ),\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'holo',\n",
       "     ),\n",
       " ),\n",
       " 'alternate_structures': {'2ajq_E': Structure(\n",
       "      (\n",
       "          'id',\n",
       "          '2ajq_E',\n",
       "      ),\n",
       "      (\n",
       "          'protein_path',\n",
       "          /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/2ajq_E.cif,\n",
       "      ),\n",
       "      (\n",
       "          'protein_sequence',\n",
       "          {\n",
       "              '1.B': 'MIVSAIAANALLESVTKFHCGVIYDYSTAEYVSYRPSDFGAYLDALEAEVARGGLIVFHNGHKYDVPALTKLAKLQLNREFHLPRENCIDTLVLSRLIHSNLKDTDMGLLRSGKLPGKRFGSHALEAWGYRLGEMKGEYKDDFKRMLEEQGEEYVDGMEWWNFNEEMMDYNVQDVVVTKALLEKLLSDKHYFPPEIDFTDVGYTTFWSESLEAVDIEHRAAWLLAKQERNGFPFDTKAIEELYVELAARRSELLRKLTETFGSWYQPKGGTEMFCHPRTGKPLPKYPRIKTPKVGGIFKKPKNKAQREGREPCELDTREYVAGAPYTPVEHVVFNPSSRDHIQKKLQEAGWVPTKYTDKGAPVVDDEVLEGVRVDDPEKQAAIDLIKEYLMIQKRIGQSAEGDKAWLRYVAEDGKIHGSVNPNGAVTGRATHAFPNLAQIPGVRSPYGEQCRAAFGAEHHLDGITGKPWVQAGIDASGLELRCLAHFMARFDNGEYAHEILNGDIHTKNQIAAELPTRDNAKTFIYGFLYGAGDEKIGQIVGAGKERGKELKKKFLENTPAIAALRESIQQTLVESSQWVAGEQQVKWKRRWIKGLDGRKVHVRSPHAALNTLLQSAGALICKLWIIKTEEMLVEKGLKHGWDGDFAYMAWVHDEIQVGCRTEEIAQVVIETAQEAMRWVGDHWNFRCLLDTEGKMGPNWAICH',\n",
       "          },\n",
       "      ),\n",
       "      (\n",
       "          'ligand_sdfs',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'ligand_smiles',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'protein_atom_array',\n",
       "          <class 'biotite.structure.AtomArray'> with shape (5541,),\n",
       "      ),\n",
       "      (\n",
       "          'ligand_mols',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'add_ligand_hydrogens',\n",
       "          False,\n",
       "      ),\n",
       "      (\n",
       "          'structure_type',\n",
       "          'apo',\n",
       "      ),\n",
       "  ),\n",
       "  '2ajq_G': Structure(\n",
       "      (\n",
       "          'id',\n",
       "          '2ajq_G',\n",
       "      ),\n",
       "      (\n",
       "          'protein_path',\n",
       "          /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/2ajq_G.cif,\n",
       "      ),\n",
       "      (\n",
       "          'protein_sequence',\n",
       "          {\n",
       "              '1.B': 'MIVSAIAANALLESVTKFHCGVIYDYSTAEYVSYRPSDFGAYLDALEAEVARGGLIVFHNGHKYDVPALTKLAKLQLNREFHLPRENCIDTLVLSRLIHSNLKDTDMGLLRSGKLPGKRFGSHALEAWGYRLGEMKGEYKDDFKRMLEEQGEEYVDGMEWWNFNEEMMDYNVQDVVVTKALLEKLLSDKHYFPPEIDFTDVGYTTFWSESLEAVDIEHRAAWLLAKQERNGFPFDTKAIEELYVELAARRSELLRKLTETFGSWYQPKGGTEMFCHPRTGKPLPKYPRIKTPKVGGIFKKPKNKAQREGREPCELDTREYVAGAPYTPVEHVVFNPSSRDHIQKKLQEAGWVPTKYTDKGAPVVDDEVLEGVRVDDPEKQAAIDLIKEYLMIQKRIGQSAEGDKAWLRYVAEDGKIHGSVNPNGAVTGRATHAFPNLAQIPGVRSPYGEQCRAAFGAEHHLDGITGKPWVQAGIDASGLELRCLAHFMARFDNGEYAHEILNGDIHTKNQIAAELPTRDNAKTFIYGFLYGAGDEKIGQIVGAGKERGKELKKKFLENTPAIAALRESIQQTLVESSQWVAGEQQVKWKRRWIKGLDGRKVHVRSPHAALNTLLQSAGALICKLWIIKTEEMLVEKGLKHGWDGDFAYMAWVHDEIQVGCRTEEIAQVVIETAQEAMRWVGDHWNFRCLLDTEGKMGPNWAICH',\n",
       "          },\n",
       "      ),\n",
       "      (\n",
       "          'ligand_sdfs',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'ligand_smiles',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'protein_atom_array',\n",
       "          <class 'biotite.structure.AtomArray'> with shape (5459,),\n",
       "      ),\n",
       "      (\n",
       "          'ligand_mols',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'add_ligand_hydrogens',\n",
       "          False,\n",
       "      ),\n",
       "      (\n",
       "          'structure_type',\n",
       "          'apo',\n",
       "      ),\n",
       "  ),\n",
       "  '5ikn_A': Structure(\n",
       "      (\n",
       "          'id',\n",
       "          '5ikn_A',\n",
       "      ),\n",
       "      (\n",
       "          'protein_path',\n",
       "          /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/5ikn_A.cif,\n",
       "      ),\n",
       "      (\n",
       "          'protein_sequence',\n",
       "          {\n",
       "              '1.B': 'MIVSAIAANALLESVTKFHCGVIYDYSTAEYVSYRPSDFGAYLDALEAEVARGGLIVFHNGHKYDVPALTKLAKLQLNREFHLPRENCIDTLVLSRLIHSNLKDTDMGLLRSGKLPGKRFGSHALEAWGYRLGEMKGEYKDDFKRMLEEQGEEYVDGMEWWNFNEEMMDYNVQDVVVTKALLEKLLSDKHYFPPEIDFTDVGYTTFWSESLEAVDIEHRAAWLLAKQERNGFPFDTKAIEELYVELAARRSELLRKLTETFGSWYQPKGGTEMFCHPRTGKPLPKYPRIKTPKVGGIFKKPKNKAQREGREPCELDTREYVAGAPYTPVEHVVFNPSSRDHIQKKLQEAGWVPTKYTDKGAPVVDDEVLEGVRVDDPEKQAAIDLIKEYLMIQKRIGQSAEGDKAWLRYVAEDGKIHGSVNPNGAVTGRATHAFPNLAQIPGVRSPYGEQCRAAFGAEHHLDGITGKPWVQAGIDASGLELRCLAHFMARFDNGEYAHEILNGDIHTKNQIAAELPTRDNAKTFIYGFLYGAGDEKIGQIVGAGKERGKELKKKFLENTPAIAALRESIQQTLVESSQWVAGEQQVKWKRRWIKGLDGRKVHVRSPHAALNTLLQSAGALICKLWIIKTEEMLVEKGLKHGWDGDFAYMAWVHDEIQVGCRTEEIAQVVIETAQEAMRWVGDHWNFRCLLDTEGKMGPNWAICH',\n",
       "          },\n",
       "      ),\n",
       "      (\n",
       "          'ligand_sdfs',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'ligand_smiles',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'protein_atom_array',\n",
       "          <class 'biotite.structure.AtomArray'> with shape (5092,),\n",
       "      ),\n",
       "      (\n",
       "          'ligand_mols',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'add_ligand_hydrogens',\n",
       "          False,\n",
       "      ),\n",
       "      (\n",
       "          'structure_type',\n",
       "          'apo',\n",
       "      ),\n",
       "  ),\n",
       "  '5ikn_B': Structure(\n",
       "      (\n",
       "          'id',\n",
       "          '5ikn_B',\n",
       "      ),\n",
       "      (\n",
       "          'protein_path',\n",
       "          /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/5ikn_B.cif,\n",
       "      ),\n",
       "      (\n",
       "          'protein_sequence',\n",
       "          {\n",
       "              '1.B': 'MIVSAIAANALLESVTKFHCGVIYDYSTAEYVSYRPSDFGAYLDALEAEVARGGLIVFHNGHKYDVPALTKLAKLQLNREFHLPRENCIDTLVLSRLIHSNLKDTDMGLLRSGKLPGKRFGSHALEAWGYRLGEMKGEYKDDFKRMLEEQGEEYVDGMEWWNFNEEMMDYNVQDVVVTKALLEKLLSDKHYFPPEIDFTDVGYTTFWSESLEAVDIEHRAAWLLAKQERNGFPFDTKAIEELYVELAARRSELLRKLTETFGSWYQPKGGTEMFCHPRTGKPLPKYPRIKTPKVGGIFKKPKNKAQREGREPCELDTREYVAGAPYTPVEHVVFNPSSRDHIQKKLQEAGWVPTKYTDKGAPVVDDEVLEGVRVDDPEKQAAIDLIKEYLMIQKRIGQSAEGDKAWLRYVAEDGKIHGSVNPNGAVTGRATHAFPNLAQIPGVRSPYGEQCRAAFGAEHHLDGITGKPWVQAGIDASGLELRCLAHFMARFDNGEYAHEILNGDIHTKNQIAAELPTRDNAKTFIYGFLYGAGDEKIGQIVGAGKERGKELKKKFLENTPAIAALRESIQQTLVESSQWVAGEQQVKWKRRWIKGLDGRKVHVRSPHAALNTLLQSAGALICKLWIIKTEEMLVEKGLKHGWDGDFAYMAWVHDEIQVGCRTEEIAQVVIETAQEAMRWVGDHWNFRCLLDTEGKMGPNWAICH',\n",
       "          },\n",
       "      ),\n",
       "      (\n",
       "          'ligand_sdfs',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'ligand_smiles',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'protein_atom_array',\n",
       "          <class 'biotite.structure.AtomArray'> with shape (5073,),\n",
       "      ),\n",
       "      (\n",
       "          'ligand_mols',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'add_ligand_hydrogens',\n",
       "          False,\n",
       "      ),\n",
       "      (\n",
       "          'structure_type',\n",
       "          'apo',\n",
       "      ),\n",
       "  ),\n",
       "  '5ikn_C': Structure(\n",
       "      (\n",
       "          'id',\n",
       "          '5ikn_C',\n",
       "      ),\n",
       "      (\n",
       "          'protein_path',\n",
       "          /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/5ikn_C.cif,\n",
       "      ),\n",
       "      (\n",
       "          'protein_sequence',\n",
       "          {\n",
       "              '1.B': 'MIVSAIAANALLESVTKFHCGVIYDYSTAEYVSYRPSDFGAYLDALEAEVARGGLIVFHNGHKYDVPALTKLAKLQLNREFHLPRENCIDTLVLSRLIHSNLKDTDMGLLRSGKLPGKRFGSHALEAWGYRLGEMKGEYKDDFKRMLEEQGEEYVDGMEWWNFNEEMMDYNVQDVVVTKALLEKLLSDKHYFPPEIDFTDVGYTTFWSESLEAVDIEHRAAWLLAKQERNGFPFDTKAIEELYVELAARRSELLRKLTETFGSWYQPKGGTEMFCHPRTGKPLPKYPRIKTPKVGGIFKKPKNKAQREGREPCELDTREYVAGAPYTPVEHVVFNPSSRDHIQKKLQEAGWVPTKYTDKGAPVVDDEVLEGVRVDDPEKQAAIDLIKEYLMIQKRIGQSAEGDKAWLRYVAEDGKIHGSVNPNGAVTGRATHAFPNLAQIPGVRSPYGEQCRAAFGAEHHLDGITGKPWVQAGIDASGLELRCLAHFMARFDNGEYAHEILNGDIHTKNQIAAELPTRDNAKTFIYGFLYGAGDEKIGQIVGAGKERGKELKKKFLENTPAIAALRESIQQTLVESSQWVAGEQQVKWKRRWIKGLDGRKVHVRSPHAALNTLLQSAGALICKLWIIKTEEMLVEKGLKHGWDGDFAYMAWVHDEIQVGCRTEEIAQVVIETAQEAMRWVGDHWNFRCLLDTEGKMGPNWAICH',\n",
       "          },\n",
       "      ),\n",
       "      (\n",
       "          'ligand_sdfs',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'ligand_smiles',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'protein_atom_array',\n",
       "          <class 'biotite.structure.AtomArray'> with shape (5041,),\n",
       "      ),\n",
       "      (\n",
       "          'ligand_mols',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'add_ligand_hydrogens',\n",
       "          False,\n",
       "      ),\n",
       "      (\n",
       "          'structure_type',\n",
       "          'apo',\n",
       "      ),\n",
       "  )},\n",
       " 'features_and_coords': {'sequence_atom_mask_feature': tensor([[1, 1, 1,  ..., 1, 1, 1]]),\n",
       "  'input_sequence_residue_mask_feature': tensor([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0.,\n",
       "           0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "           1., 1.]], dtype=torch.float64),\n",
       "  'protein_coordinates': tensor([[[  52.5260, -103.6860,   94.7630],\n",
       "           [  53.2660, -104.2480,   93.6390],\n",
       "           [  52.3430, -105.0080,   92.6970],\n",
       "           ...,\n",
       "           [  88.9380,  -94.2960,   71.2860],\n",
       "           [  89.1790,  -94.7920,   70.0850],\n",
       "           [  90.8010,  -91.0330,   69.1100]]]),\n",
       "  'protein_calpha_coordinates': tensor([[[  53.2660, -104.2480,   93.6390],\n",
       "           [  51.7380, -105.4890,   90.3850],\n",
       "           [  52.9920, -107.7890,   87.6240],\n",
       "           ...,\n",
       "           [  87.2200,  -86.3160,   65.8780],\n",
       "           [  86.0480,  -87.3630,   69.3720],\n",
       "           [  88.5370,  -90.2720,   69.5410]]]),\n",
       "  'input_sequence_full_atom_feature': tensor([[[0., 0., 0.,  ..., 0., 0., 0.],\n",
       "           [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "           [1., 0., 0.,  ..., 0., 0., 0.],\n",
       "           ...,\n",
       "           [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "           [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "           [0., 0., 0.,  ..., 0., 0., 0.]]], dtype=torch.float64),\n",
       "  'protein_structure_residue_feature': tensor([[[0., 0., 0.,  ..., 0., 0., 0.],\n",
       "           [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "           [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "           ...,\n",
       "           [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "           [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "           [0., 0., 0.,  ..., 0., 0., 0.]]], dtype=torch.float64),\n",
       "  'input_conformer_ligand_feature': tensor([[[ 5,  0,  4,  5,  3,  3,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 5,  0,  3,  5,  0,  0,  0,  1,  1,  1,  0,  0,  0,  1,  0,  0],\n",
       "           [ 5,  0,  3,  5,  1,  1,  0,  1,  1,  1,  0,  0,  0,  1,  0,  0],\n",
       "           [ 6,  0,  3,  5,  0,  0,  0,  1,  1,  1,  0,  0,  0,  1,  0,  0],\n",
       "           [ 5,  1,  4,  5,  0,  1,  0,  2,  0,  1,  0,  0,  1,  0,  0,  0],\n",
       "           [ 5,  0,  4,  5,  2,  2,  0,  2,  0,  1,  0,  0,  1,  0,  0,  0],\n",
       "           [ 5,  2,  4,  5,  0,  1,  0,  2,  0,  1,  0,  0,  1,  0,  0,  0],\n",
       "           [ 7,  0,  2,  5,  1,  1,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 5,  1,  4,  5,  0,  1,  0,  2,  0,  1,  0,  0,  1,  0,  0,  0],\n",
       "           [ 5,  0,  4,  5,  2,  2,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  2,  5,  0,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [14,  2,  4,  5,  0,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  1,  5,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  2,  5,  1,  1,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  2,  5,  0,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [14,  1,  4,  5,  0,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  1,  5,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  2,  5,  1,  1,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  2,  5,  0,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [14,  0,  4,  5,  0,  0,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  1,  5,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  2,  5,  1,  1,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  2,  5,  1,  1,  0,  2,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 7,  0,  2,  5,  0,  0,  0,  2,  0,  1,  0,  0,  1,  0,  0,  0],\n",
       "           [ 5,  0,  3,  5,  0,  0,  0,  1,  1,  1,  0,  0,  0,  1,  0,  0],\n",
       "           [ 7,  0,  1,  5,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0],\n",
       "           [ 6,  0,  3,  5,  0,  1,  0,  1,  1,  1,  0,  0,  0,  1,  0,  0],\n",
       "           [ 5,  0,  3,  5,  0,  0,  0,  1,  1,  1,  0,  0,  0,  1,  0,  0],\n",
       "           [ 7,  0,  1,  5,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0]]]),\n",
       "  'input_conformer_ligand_coordinates': tensor([[[-5.3903e+00, -2.0348e+00, -1.3399e+00],\n",
       "           [-4.8322e+00, -6.6411e-01, -1.1445e+00],\n",
       "           [-3.5828e+00, -4.0656e-01, -7.3874e-01],\n",
       "           [-3.1024e+00,  8.7929e-01, -5.7661e-01],\n",
       "           [-1.7499e+00,  1.1314e+00, -9.1964e-02],\n",
       "           [-1.6924e+00,  1.3608e+00,  1.4090e+00],\n",
       "           [-2.4874e-01,  1.0529e+00,  1.7260e+00],\n",
       "           [ 5.2589e-01,  2.2482e+00,  1.6318e+00],\n",
       "           [ 1.7731e-01,  6.8044e-02,  6.3230e-01],\n",
       "           [ 4.5615e-01, -1.3259e+00,  1.1924e+00],\n",
       "           [ 1.1041e+00, -2.1416e+00,  2.2573e-01],\n",
       "           [ 2.6998e+00, -2.2862e+00,  2.0504e-01],\n",
       "           [ 3.2027e+00, -3.3694e+00, -6.9340e-01],\n",
       "           [ 3.0815e+00, -2.6220e+00,  1.7203e+00],\n",
       "           [ 3.3302e+00, -8.6701e-01, -9.5770e-02],\n",
       "           [ 3.8563e+00,  2.4027e-01,  8.9658e-01],\n",
       "           [ 2.8768e+00,  7.1829e-01,  1.9253e+00],\n",
       "           [ 5.2219e+00, -3.4521e-01,  1.4929e+00],\n",
       "           [ 4.3791e+00,  1.4247e+00, -2.9613e-02],\n",
       "           [ 5.2202e+00,  1.2104e+00, -1.3498e+00],\n",
       "           [ 4.4975e+00,  7.9754e-01, -2.5786e+00],\n",
       "           [ 6.0395e+00,  2.5690e+00, -1.5073e+00],\n",
       "           [ 6.3773e+00,  1.9565e-01, -8.9938e-01],\n",
       "           [-8.9733e-01, -1.5862e-03, -3.2857e-01],\n",
       "           [-3.9060e+00,  1.9879e+00, -8.0222e-01],\n",
       "           [-3.5466e+00,  3.1500e+00, -6.2645e-01],\n",
       "           [-5.1760e+00,  1.7214e+00, -1.2384e+00],\n",
       "           [-5.7247e+00,  4.8542e-01, -1.4183e+00],\n",
       "           [-6.8877e+00,  3.3430e-01, -1.7762e+00]]], dtype=torch.float64),\n",
       "  'resolved_ligand_mols_feature': tensor([[[  60.2230, -111.8430,   70.4970],\n",
       "           [  60.9010, -110.6400,   69.8680],\n",
       "           [  61.1870, -112.5560,   71.4190],\n",
       "           [  59.6040, -112.8290,   69.3780],\n",
       "           [  59.2050, -114.3600,   69.7240],\n",
       "           [  57.9260, -114.7480,   69.0100],\n",
       "           [  59.0170, -114.5670,   71.2070],\n",
       "           [  60.4600, -115.2160,   69.1670],\n",
       "           [  61.8790, -115.2600,   69.9490],\n",
       "           [  62.6810, -114.0100,   69.6620],\n",
       "           [  61.6710, -115.3820,   71.4410],\n",
       "           [  62.6790, -116.4520,   69.4770],\n",
       "           [  58.9280, -111.4030,   71.3430],\n",
       "           [  58.8750, -111.8610,   72.6880],\n",
       "           [  57.6870, -111.3110,   73.4710],\n",
       "           [  57.3410, -109.9770,   73.0730],\n",
       "           [  56.4300, -112.1480,   73.3080],\n",
       "           [  56.1580, -112.8580,   74.5250],\n",
       "           [  55.3170, -111.1550,   72.9930],\n",
       "           [  55.9530, -109.7930,   72.7530],\n",
       "           [  55.6900, -109.4190,   71.3560],\n",
       "           [  54.4260, -108.8550,   71.0740],\n",
       "           [  53.6360, -108.6320,   72.0130],\n",
       "           [  54.0750, -108.4840,   69.8390],\n",
       "           [  54.8960, -108.6900,   68.8070],\n",
       "           [  54.5630, -108.3570,   67.6510],\n",
       "           [  56.2140, -109.3190,   69.0550],\n",
       "           [  57.1730, -109.5640,   67.9210],\n",
       "           [  56.5460, -109.6750,   70.3580]]], dtype=torch.float64)},\n",
       " 'path': '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/6p7e__2__1.B__1.Y/system.cif'}"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_vata_data_point"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "test_data[110]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_loader = get_torch_loader(\n",
    "    val_dataset\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 13:16:06,544 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:06,544 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:07,347 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.66s\n",
      "2024-09-23 13:16:07,476 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.13s\n",
      "2024-09-23 13:16:07,476 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-23 13:16:07,482 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-23 13:16:07,482 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 0.93s\n",
      "2024-09-23 13:16:07,520 | plinder.core.structure.atoms:207 | WARNING : get_template_to_mol_matches: could not match template fully - retry with unmatched bonds set as UNSPECIFIED\n",
      "2024-09-23 13:16:07,529 | plinder.core.structure.atoms:207 | WARNING : get_template_to_mol_matches: could not match template fully - retry with unmatched bonds set as UNSPECIFIED\n",
      "2024-09-23 13:16:08,020 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.18s\n",
      "2024-09-23 13:16:08,185 | plinder.core.scores.links.query_links:24 | INFO : runtime succeeded: 0.57s\n",
      "2024-09-23 13:16:12,920 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 4.60s\n",
      "2024-09-23 13:16:13,056 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.14s\n",
      "2024-09-23 13:16:18,446 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.65s\n",
      "2024-09-23 13:16:18,568 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.12s\n",
      "2024-09-23 13:16:18,568 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-23 13:16:18,620 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-23 13:16:18,620 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 0.95s\n",
      "2024-09-23 13:16:18,722 | plinder.core.structure.atoms:121 | INFO : generate_conformer: MMFFOptimizeMolecule - more iterations are required, extending by 500 steps\n",
      "2024-09-23 13:16:19,114 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.14s\n",
      "2024-09-23 13:16:19,386 | plinder.core.scores.links.query_links:24 | INFO : runtime succeeded: 0.62s\n",
      "2024-09-23 13:16:19,752 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:19,752 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:20,055 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:20,055 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:21,180 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.72s\n",
      "2024-09-23 13:16:21,269 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.09s\n",
      "2024-09-23 13:16:21,269 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-23 13:16:21,271 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-23 13:16:21,271 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 1.21s\n",
      "2024-09-23 13:16:21,731 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.14s\n",
      "2024-09-23 13:16:22,218 | plinder.core.scores.links.query_links:24 | INFO : runtime succeeded: 0.77s\n",
      "2024-09-23 13:16:22,608 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:22,608 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:22,874 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:22,874 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:23,598 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.59s\n",
      "2024-09-23 13:16:23,729 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.13s\n",
      "2024-09-23 13:16:23,729 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-23 13:16:23,734 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-23 13:16:23,734 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 0.86s\n",
      "2024-09-23 13:16:24,183 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.14s\n",
      "2024-09-23 13:16:24,436 | plinder.core.scores.links.query_links:24 | INFO : runtime succeeded: 0.58s\n",
      "2024-09-23 13:16:24,859 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 13:16:24,859 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n"
     ]
    }
   ],
   "source": [
    "for data in train_loader:\n",
    "\n",
    "    test_torch = data\n",
    "    break\n",
    "    #for k, v in test_torch['input_features'].items():\n",
    "    #    if v.shape[1] > 1:\n",
    "    #        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['system_ids', 'holo_structures', 'alternate_structures', 'paths', 'features_and_coords'])"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_torch.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['3dug__1__1.E__1.IA_1.LA_1.MA', '2iyg__1__2.B__2.E']"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_torch['system_ids']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence_atom_mask_feature torch.Size([2, 1, 3269])\n",
      "input_sequence_residue_mask_feature torch.Size([2, 1, 408])\n",
      "protein_coordinates torch.Size([2, 1, 3001, 3])\n",
      "protein_calpha_coordinates torch.Size([2, 1, 394, 3])\n",
      "input_sequence_full_atom_feature torch.Size([2, 1, 3115, 12])\n",
      "protein_structure_residue_feature torch.Size([2, 1, 3001, 21])\n",
      "input_conformer_ligand_feature torch.Size([2, 3, 31, 16])\n",
      "input_conformer_ligand_coordinates torch.Size([2, 3, 31, 3])\n",
      "resolved_ligand_mols_feature torch.Size([2, 3, 31, 3])\n"
     ]
    }
   ],
   "source": [
    "for k, v in test_torch['features_and_coords'].items():\n",
    "    print(k, v.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.C': (<rdkit.Chem.rdchem.Mol at 0x15f2c2110>,\n",
       "  <rdkit.Chem.rdchem.Mol at 0x15f2c26c0>,\n",
       "  <rdkit.Chem.rdchem.Mol at 0x15f2c1380>,\n",
       "  (array([[13,  4,  5,  7,  9, 10,  1,  0,  3,  6,  8, 12, 11,  2]]),\n",
       "   array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13]])))}"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.ligand_mols"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Align and superpose structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alternate_structure = sample_system.alternate_structures['1vyo_B']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 12:42:47,142 | plinder.core.structure.atoms:121 | INFO : generate_conformer: MMFFOptimizeMolecule - more iterations are required, extending by 500 steps\n"
     ]
    }
   ],
   "source": [
    "holo_struc = sample_system.holo_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alternate_structure.set_chain(list(alternate_structure.protein_sequence.keys())[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alignment = holo_struc.align_common_sequence(alternate_structure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Structure(\n",
       "    (\n",
       "        'id',\n",
       "        '1avd__1__1.A__1.C',\n",
       "    ),\n",
       "    (\n",
       "        'protein_path',\n",
       "        /Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.cif,\n",
       "    ),\n",
       "    (\n",
       "        'protein_sequence',\n",
       "        {\n",
       "            '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_sdfs',\n",
       "        {\n",
       "            '1.C': '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_smiles',\n",
       "        {\n",
       "            '1.C': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'protein_atom_array',\n",
       "        <class 'biotite.structure.AtomArray'> with shape (930,),\n",
       "    ),\n",
       "    (\n",
       "        'ligand_mols',\n",
       "        {\n",
       "            '1.C': (\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x1bf9fb760>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x1bf9fb4c0>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x1bf9fb7d0>,\n",
       "                (\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                ),\n",
       "            ),\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'add_ligand_hydrogens',\n",
       "        False,\n",
       "    ),\n",
       "    (\n",
       "        'structure_type',\n",
       "        'holo',\n",
       "    ),\n",
       ")"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert alignment[0].protein_unique_residue_ids == alignment[1].protein_unique_residue_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "superimposed_struc = alignment[0].superimpose(alignment[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1avd__1__1.A__1.C',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         {\n",
       "             '1.C': '/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         {\n",
       "             '1.C': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (930,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         {\n",
       "             '1.C': (\n",
       "                 <rdkit.Chem.rdchem.Mol object at 0x14a17f680>,\n",
       "                 <rdkit.Chem.rdchem.Mol object at 0x14a17f1b0>,\n",
       "                 <rdkit.Chem.rdchem.Mol object at 0x14a17f5a0>,\n",
       "                 (\n",
       "                     <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                     <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                 ),\n",
       "             ),\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'holo',\n",
       "     ),\n",
       " ),\n",
       " 0.45251316,\n",
       " 0.28305227)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "superimposed_struc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
